<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Normalize CITEseq Data • dsb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Normalize CITEseq Data">
<meta property="og:description" content="see biorxiv preprint DOI:Here we developed a method specifically for normalizing CITEseq data,accounting for protein-specific background noise caused by unbound antibody captured and sequenced in droplets as well as correcting for the technical component of variation in protein library size. We account for these sources of experimental noisewith the empty droplets containing background antibody reads and by defining a covariate of non-staining proteins and isotype control antibodies respectively.">
<meta property="og:image" content="https://mattpm.github.io/dsb/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">dsb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/dsb_normalizing_CITEseq_data.html">dsb_normalizing_CITEseq_data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/dsb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="dsb-an-r-package-for-normalizing-and-denoising-citeseq-protein-data" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#dsb-an-r-package-for-normalizing-and-denoising-citeseq-protein-data" class="anchor"></a>dsb : an R package for normalizing and denoising CITEseq protein data <a href="https://mattpm.github.io/dsb"><img src="reference/figures/logo.png" align="right" height="150"></a>
</h1></div>
<p>Package maintainer: Matt Mulè<br>
Package hosted by: National Institutes of Allergy and Infectious Diseases, NIH</p>
<!-- badges: start -->
<!-- [![Travis build status](https://travis-ci.org/MattPM/dsb.svg?branch=master)](https://travis-ci.org/MattPM/dsb) -->
<!-- badges: end -->
<p>dsb (<em>d</em>enoised and <em>s</em>caled by <em>b</em>ackground) is a lightweight R package for normalizing and denoising protein expression data from CITE-seq. The dsb package was developed in <a href="https://www.niaid.nih.gov/research/john-tsang-phd">John Tsang’s Lab at the NIH-NIAID</a> by Matt Mulè, Andrew Martins and John Tsang. If you use dsb or find the noise modeling results in our paper useful in your research please consider citing the preprint below.</p>
<p><a href="https://www.biorxiv.org/content/10.1101/2020.02.24.963603v1.full.pdf">our biorxiv preprint</a> provides details of the dsb method.<br><a href="https://doi.org/10.1038/s41591-020-0769-8">our Nature Medicine paper</a> shows the utility of dsb normalized values for CITE-seq clustering, modeling and visualization</p>
<p>The method is implemented through a single function call to DSBNormalizeProtein which operates on 2 matrices of raw protein counts: cells x protein and empty/background droplets x protein. Our method is based on experiments and analysis of several datasets including control experiments designed to interrogate the source of background noise in CITE-seq experiments in which we discovered:</p>
<p>1) a major source of protein-specific noise in CITE-seq data originates from ambient, unbound antibody encapsulated in droplets, which we correct for by estimating the expected protein counts detected in empty droplets.</p>
<p>2) cell-to-cell technical noise can be estimated and removed via the shared variance component associated with isotype antibody controls and background protein counts in each cell.</p>
<p>Below we outline the DSB method starting from raw Cell Ranger output from a public 10X CITE-seq dataset. We estimate empty/background and cell-containing droplets, and normalize protein with the DSBNormalizeProtein function which returns a normalized protein data that can be integrated with <em>any single cell analysis software</em>. We also show a suggested workflow for protein based clustering directly on DSB Normalized values, dimensionality reduction, and visualization using R and Seurat.</p>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/dsb_umap.png"></a></p>
<div id="installation-and-quick-overview-with-pre-loaded-package-data" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-and-quick-overview-with-pre-loaded-package-data" class="anchor"></a>Installation and quick overview with pre-loaded package data</h2>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/dsb_overview.png"></a></p>
<div class="sourceCode" id="cb1"><pre class="r"><span class="co"># install dsb package directly from github: </span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">devtools</span>); <span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span>(<span class="kw">repo</span> <span class="kw">=</span> <span class="st">'MattPM/dsb'</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dsb</span>)

<span class="co"># define isotype controls </span>
<span class="no">isotypes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">cells_citeseq_mtx</span>)[<span class="fl">67</span>:<span class="fl">70</span>]
<span class="co"># cells_citeseq_mtx = cells with raw protein counts (loaded with dsb)</span>
<span class="co"># empty_drop_citeseq_mtx = empty/background droplets with raw protein counts(loaded with dsb)</span>

<span class="no">adt_norm</span> <span class="kw">=</span> <span class="fu"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(
  <span class="co"># Input = two raw CITE-seq protein count matrices 1) cell-containing droplets, 2) empty droplets</span>
  <span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">cells_citeseq_mtx</span>,
  <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">empty_drop_citeseq_mtx</span>,

  <span class="co"># Step II arguments </span>
  <span class="kw">denoise.counts</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="co"># Recommended - denoise each cell's background rescaled values from step I above</span>
  <span class="kw">use.isotype.control</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="co"># whether to utilize isotype controls to define the denoising covariate (highly recommended)</span>
  <span class="kw">isotype.control.name.vec</span> <span class="kw">=</span> <span class="no">isotypes</span> <span class="co"># a vector with names of isotype control proteins (see above) </span>
  )</pre></div>
</div>
<div id="tutorial-step-i-dsb-normalize-cite-seq-protein-data-starting-from-raw-cell-ranger-output" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorial-step-i-dsb-normalize-cite-seq-protein-data-starting-from-raw-cell-ranger-output" class="anchor"></a>Tutorial Step I: DSB normalize CITE-seq protein data starting from <em>RAW</em> cell ranger output</h2>
<p>Below, raw protein UMI counts for empty droplets and cells are estimated using the protein and mRNA library size distributions directly from the cell ranger (raw) output. These are used as above to normalize the protein data for cells using the dsb package. Normalized protein data and raw RNA data can then be used to create a Seurat object, Bioconductor’s SingleCellExperiment object, or an AnnData object in python for downstream analysis.</p>
<p>To follow this example download <strong>RAW (not filtered!)</strong> <strong>feature / cellmatrix raw</strong> public 10X CITE-seq data from here: <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_protein_v3" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_protein_v3</a>. The steps below use R 3.6 and Seurat version 3.</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyverse</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dsb</span>)

<span class="co"># read raw data using the Seurat function "Read10X"</span>
<span class="no">raw</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="st">"data/10x_data/10x_pbmc5k_V3/raw_feature_bc_matrix/"</span>)

<span class="co"># Define separate RNA and protein sparse matrix  </span>
<span class="no">prot</span> <span class="kw">=</span> <span class="no">raw</span>$<span class="no">`Antibody Capture`</span>
<span class="no">rna</span> <span class="kw">=</span> <span class="no">raw</span>$<span class="no">`Gene Expression`</span>

<span class="co"># calculate metadata </span>
<span class="no">rna_size</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">rna</span>))
<span class="no">prot_size</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">prot</span>))
<span class="no">ngene</span> <span class="kw">=</span> <span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">rna</span> <span class="kw">&gt;</span> <span class="fl">0</span>)
<span class="no">mtgene</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span>(<span class="kw">pattern</span> <span class="kw">=</span> <span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">rna</span>), <span class="kw">value</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">propmt</span> <span class="kw">=</span> <span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">rna</span>[<span class="no">mtgene</span>, ]) / <span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">rna</span>)
<span class="no">md</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">propmt</span>, <span class="no">rna_size</span>, <span class="no">ngene</span>, <span class="no">prot_size</span>))
<span class="no">md</span>$<span class="no">bc</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">md</span>)

<span class="co"># histogram to estimate cells and background-cells are a tiny fraction of drops with log 10 protien lib size &gt; 3</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="no">md</span>$<span class="no">log10umiprot</span>[<span class="no">md</span>$<span class="no">log10umiprot</span> <span class="kw">&lt;</span> <span class="fl">5</span>], <span class="kw">breaks</span> <span class="kw">=</span> <span class="fl">100</span>)

<span class="co"># define a vector of background / empty droplet barcodes based on protein library size and mRNA content  </span>
<span class="no">background_drops</span> <span class="kw">=</span> <span class="no">md</span>[<span class="no">md</span>$<span class="no">prot_size</span> <span class="kw">&lt;</span> <span class="fl">2.5</span> <span class="kw">&amp;</span> <span class="no">md</span>$<span class="no">prot_size</span> <span class="kw">&gt;</span> <span class="fl">1.4</span> <span class="kw">&amp;</span> <span class="no">md</span>$<span class="no">ngene</span> <span class="kw">&lt;</span> <span class="fl">80</span>, ]$<span class="no">bc</span>
<span class="no">negative_mtx_rawprot</span> <span class="kw">=</span> <span class="no">prot</span>[ , <span class="no">background_drops</span>] <span class="kw">%&gt;%</span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()

<span class="co"># define a vector of cell-containing droplet barcodes based on protein library size and mRNA content </span>
<span class="no">positive_cells</span> <span class="kw">=</span> <span class="no">md</span>[<span class="no">md</span>$<span class="no">prot_size</span> <span class="kw">&gt;</span> <span class="fl">2.8</span> <span class="kw">&amp;</span> <span class="no">md</span>$<span class="no">ngene</span> <span class="kw">&lt;</span> <span class="fl">3000</span> <span class="kw">&amp;</span> <span class="no">md</span>$<span class="no">ngene</span> <span class="kw">&gt;</span> <span class="fl">200</span> <span class="kw">&amp;</span> <span class="no">propmt</span> <span class="kw">&lt;</span><span class="fl">0.2</span>, ]$<span class="no">bc</span>
<span class="no">cells_mtx_rawprot</span> <span class="kw">=</span> <span class="no">prot</span>[ , <span class="no">positive_cells</span>] <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()

<span class="co">#normalize protein data for the cell containing droplets with the dsb method. </span>
<span class="no">isotypes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">pos_prot</span>)[<span class="fl">30</span>:<span class="fl">32</span>]
<span class="no">dsb_norm_prot</span> <span class="kw">=</span> <span class="fu"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(
                           <span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">cells_mtx_rawprot</span>,
                           <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">negative_mtx_rawprot</span>,
                           <span class="kw">denoise.counts</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                           <span class="kw">use.isotype.control</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                           <span class="kw">isotype.control.name.vec</span> <span class="kw">=</span> <span class="no">isotypes</span>
                           )</pre></div>
<p>The DSB normalized protein matrix can be integrated with any single cell analysis software, we provide an example with Seurat below.</p>
</div>
<div id="step-ii-integration-with-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#step-ii-integration-with-seurat" class="anchor"></a>Step II: integration with Seurat</h2>
<div class="sourceCode" id="cb3"><pre class="r"><span class="co"># Create Seurat object filter raw protein, RNA and metadata to only include cell-containing droplets </span>
<span class="no">count_rna</span> <span class="kw">=</span> <span class="no">raw</span>$<span class="no">`Gene Expression`</span>[ ,<span class="no">positive_cells</span>]
<span class="no">count_prot</span> <span class="kw">=</span> <span class="no">raw</span>$<span class="no">`Antibody Capture`</span>[ ,<span class="no">positive_cells</span>]

<span class="no">md</span> <span class="kw">=</span> <span class="no">md</span> <span class="kw">%&gt;%</span>
  <span class="fu">rownames_to_column</span>(<span class="st">"bc"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="no">bc</span> <span class="kw">%in%</span> <span class="no">positive_cells</span>) <span class="kw">%&gt;%</span>
  <span class="fu">column_to_rownames</span>(<span class="st">'bc'</span>)

<span class="co"># create Seurat object * note min.cells ins a gene filter not a cell filter, we alerady filtered cells in steps above</span>
<span class="no">s</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">count_rna</span>, <span class="kw">meta.data</span> <span class="kw">=</span> <span class="no">md</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"RNA"</span>, <span class="kw">min.cells</span> <span class="kw">=</span> <span class="fl">20</span>)

<span class="co"># add DSB normalized "dsb_norm_prot" protein data to the seurat object </span>
<span class="no">s</span><span class="kw">[[</span><span class="st">"CITE"</span>]] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateAssayObject.html">CreateAssayObject</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">dsb_norm_prot</span>)</pre></div>
</div>
<div id="suggested-step-iii-protein-based-clustering-cluster-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#suggested-step-iii-protein-based-clustering-cluster-annotation" class="anchor"></a>Suggested Step III: Protein based clustering + cluster annotation</h2>
<ul>
<li>This is similar to the workflow used in our paper <a href="https://www.nature.com/articles/s41591-020-0769-8" class="uri">https://www.nature.com/articles/s41591-020-0769-8</a> where we analyzed mRNA states <em>within</em> the interpretable clusters defined by clustering on DSB normalized protein data (see next section).</li>
</ul>
<!-- end list --><div class="sourceCode" id="cb4"><pre class="r"><span class="co"># define Euclidean distance matrix on dsb normalized protein data (without isotype controls)</span>
<span class="no">dsb</span> <span class="kw">=</span> <span class="no">s</span>@<span class="kw">assays</span>$<span class="no">CITE</span>@<span class="kw">data</span>[<span class="fl">1</span>:<span class="fl">29</span>, ]
<span class="no">p_dist</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">dsb</span>))
<span class="no">p_dist</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">p_dist</span>)

<span class="co"># Cluster using Seurat </span>
<span class="no">s</span><span class="kw">[[</span><span class="st">"p_dist"</span>]] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindNeighbors.html">FindNeighbors</a></span>(<span class="no">p_dist</span>)$<span class="no">snn</span>
<span class="no">s</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(<span class="no">s</span>, <span class="kw">resolution</span> <span class="kw">=</span> <span class="fl">0.5</span>, <span class="kw">graph.name</span> <span class="kw">=</span> <span class="st">"p_dist"</span>)</pre></div>
</div>
<div id="suggested-step-iv-cluster-annotation-based-on-dsb-normalized-protein-values" class="section level2">
<h2 class="hasAnchor">
<a href="#suggested-step-iv-cluster-annotation-based-on-dsb-normalized-protein-values" class="anchor"></a>Suggested Step IV: cluster annotation based on DSB normalized protein values</h2>
<p>DSB normalized values provide a straightforward comparable value for each protein in each cluster. They are the log number of standard deviations (+ / - the cell-intrinsic denoised technical component) from the expected noise as reflected by the protein distribution in empty droplets.</p>
<div class="sourceCode" id="cb5"><pre class="r">
<span class="no">prots</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">s</span>@<span class="kw">assays</span>$<span class="no">CITE</span>@<span class="kw">data</span>)
<span class="no">adt_plot</span> <span class="kw">=</span> <span class="no">adt_data</span> <span class="kw">%&gt;%</span>
  <span class="fu">group_by</span>(<span class="no">seurat_clusters</span>) <span class="kw">%&gt;%</span>
  <span class="fu">summarize_at</span>(<span class="kw">.vars</span> <span class="kw">=</span> <span class="no">prots</span>, <span class="kw">.funs</span> <span class="kw">=</span> <span class="no">mean</span>) <span class="kw">%&gt;%</span>
  <span class="fu">column_to_rownames</span>(<span class="st">"seurat_clusters"</span>) <span class="kw">%&gt;%</span>
  <span class="no">t</span> <span class="kw">%&gt;%</span>
  <span class="no">as.data.frame</span>

<span class="kw pkg">pheatmap</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span>(<span class="no">adt_plot</span>, <span class="kw">color</span> <span class="kw">=</span> <span class="kw pkg">viridis</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/viridis/man/reexports.html">viridis</a></span>(<span class="fl">25</span>, <span class="kw">option</span> <span class="kw">=</span> <span class="st">"B"</span>),
                   <span class="kw">fontsize_row</span> <span class="kw">=</span> <span class="fl">8</span>, <span class="kw">border_color</span> <span class="kw">=</span> <span class="fl">NA</span>, <span class="kw">width</span> <span class="kw">=</span> <span class="fl">5</span>, <span class="kw">height</span> <span class="kw">=</span> <span class="fl">5</span> )</pre></div>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/dsb_heatmap.png" height="500"></a></p>
<div id="calculate-umap-embedding-on-dsb-normalized-protein-values-and-visualize-all-proteins-in-single-cells-on-the-same-interpretable-dsb-scale" class="section level3">
<h3 class="hasAnchor">
<a href="#calculate-umap-embedding-on-dsb-normalized-protein-values-and-visualize-all-proteins-in-single-cells-on-the-same-interpretable-dsb-scale" class="anchor"></a>calculate UMAP embedding on DSB normalized protein values and visualize all proteins in single cells on the same interpretable DSB scale</h3>
<p>One can also calculate UMAP embeddings and to visualize the interpretable scale on which DSB places single cell protein expression levels. Note it is also possible to calculate UMAP embeddings directly with <em>Seurat</em>, <em>scater</em> etc. below we use the python umap package called through reticulate</p>
<div class="sourceCode" id="cb6"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">reticulate</span>); <span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_virtualenv</a></span>(<span class="st">"r-reticulate"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">umap</span>)

<span class="co"># set umap config</span>
<span class="no">config</span> <span class="kw">=</span> <span class="no">umap.defaults</span>
<span class="no">config</span>$<span class="no">n_neighbors</span> <span class="kw">=</span> <span class="fl">40</span>
<span class="no">config</span>$<span class="no">min_dist</span> <span class="kw">=</span> <span class="fl">0.4</span>

<span class="co"># run umap</span>
<span class="no">ump</span> <span class="kw">=</span> <span class="fu">umap</span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">s2_adt3</span>), <span class="kw">config</span> <span class="kw">=</span> <span class="no">config</span>)
<span class="no">umap_res</span> <span class="kw">=</span> <span class="no">ump</span>$<span class="no">layout</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>()
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">umap_res</span>) <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>)

<span class="co"># save results dataframe </span>
<span class="no">df_dsb</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="no">s</span>@<span class="kw">meta.data</span>, <span class="no">umap_res</span>, <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span>(<span class="no">s</span>@<span class="kw">assay</span>$<span class="no">CITE</span>@<span class="kw">data</span>)))
<span class="co"># visualizatons below were made directly from the data frame df_dsb above with ggplot</span></pre></div>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/dsb_umap.png"></a></p>
</div>
<div id="modeling-assumptions-in-background-rescaling-and-cell-denoising-in-10x-data" class="section level3">
<h3 class="hasAnchor">
<a href="#modeling-assumptions-in-background-rescaling-and-cell-denoising-in-10x-data" class="anchor"></a>modeling assumptions in background rescaling and cell denoising in 10X data</h3>
<p>In this public dataset, the number of estimated cell-free droplets (&gt;60,000 drops) with protein provided a robust estimate of the distribution of each proteins expected noise; this is consistent with our original report discovering ambient antibody as the major noise source by comparing unstained control cells and empty droplets. Part of defining the defining each cell’s technical component, is fitting a Gaussian mixture model to each cell under the assumption cells will have a bimodal staining distribution of positive and negative proteins. The background mean from a Gaussian mixture model with k=2 components globally had the best fit across single cells comparing 1,2 and 3 component models, consistent with this assumption. The estimated µ1 from the k=2 model corelated the isotype control mean and individual isotype contorls. The inferred technical component using isotype controls and µ1 had a positive non-linear relationship with the protein library size across cell clusters, suggesting this technical component controls for stochastic technical differences in library sizes between cells, but not in the 1:1 linear proportion that would be enforced by library size scaling factors (as are commonly used for mRNA normalization but are not appropriate for protein due to expected biological variation in library sizes).</p>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/10x_data_summary.png" height="500"></a></p>
</div>
</div>
<div id="more-information-how-were-background-drops-defined-in-the-quick-example-above" class="section level2">
<h2 class="hasAnchor">
<a href="#more-information-how-were-background-drops-defined-in-the-quick-example-above" class="anchor"></a>More information: How were background drops defined in the quick example above?</h2>
<p>The raw cell ranger output that we loaded contained all possible barcode sequences based on the 10x genomics Version 3 chemistry most have no mRNA or protein data, a clear peak based on log protein library size just under 2 were used to define the background for DSB normalization (see thresholds used in section I) with an additional filter to remove droplets with more than 80 unique mRNA. Cells are a tiny fraction of total droplets with library sizes &gt; 3 (matching the expected number of cells based on loading). These are not hard thresholds and can be calculated with quantile statistics inherent to each dataset depending on tissue profiled or assay chemistry used.</p>
<p>The full protein library size distribution:</p>
<div class="sourceCode" id="cb7"><pre class="r"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">rna</span>))[<span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(<span class="kw pkg">Matrix</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/colSums.html">colSums</a></span>(<span class="no">rna</span>)) <span class="kw">&gt;</span> <span class="fl">0</span>],<span class="kw">breaks</span> <span class="kw">=</span> <span class="fl">50</span>,
     <span class="kw">main</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">rna</span>),<span class="st">" droplets protein library size distribution "</span>))</pre></div>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/full_lib.png"></a></p>
<p>The library size distribution of cells and empty droplets after QC in part I above (the steps were used above to define negative and poitive droplets) are shown below with the number of each population shown-this gives a sense of the large number of barcodes with ambient protein detected (&gt; 60,000 empty droplets containing antibody measurements from which to estimate noise) vs cells ~ 3800 in line with what we expect from the experiment loading and removal of some cells during QC.</p>
<p><a href="https://mattpm.github.io/dsb"><img src="reference/figures/library_size_10x_dsb_distribution.png" height="500"></a></p>
<p>(code used to create plot above shown at the end of this markdown for reference)</p>
</div>
<div id="quickstart-v2-dsb-rescaled" class="section level2">
<h2 class="hasAnchor">
<a href="#quickstart-v2-dsb-rescaled" class="anchor"></a>Quickstart V2 dsb-rescaled</h2>
<p>If one does not have isotype control data, we recommend not defining the technical component of each cell’s protein library and not denoising counts, simply using the dsb rescaled matrix. The background mean for each cell inferred via a per-cell gaussian mixture model can alternatively be used alone to define the cell’s technical component, but this assumes the background mean has no expected biological variation. In our analysis, across single cells the background mean has weak but significant correlation with the foreground mean, calculating the eigenvector (i.e. the primary latent component) through the background mean and isotype controls and using that as the technical component of each cell’s protein library anchors the technical component of the background mean (this is what was done above and is the default method used in DSB).</p>
<div class="sourceCode" id="cb8"><pre class="r"><span class="no">dsb_rescaled</span> <span class="kw">=</span> <span class="fu"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">cells_citeseq_mtx</span>,
                                   <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">empty_drop_citeseq_mtx</span>,
                                   <span class="co"># only run dsb-rescaled do not denoise each cell's technical component</span>
                                   <span class="kw">denoise.counts</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></div>
</div>
<div id="how-else-can-backgroundempty-droplets-be-defined" class="section level2">
<h2 class="hasAnchor">
<a href="#how-else-can-backgroundempty-droplets-be-defined" class="anchor"></a>How else can background/empty droplets be defined?</h2>
<p>1) As shown above one can define background and cells with library size distributions. 2) If experiment is sample multiplexed with barcoding (i.e. hashing, multiseq, demuxlet), demultiplexing functions define a “negative” cell population which can be used to define background. In our data, the resulting DSB normalized values were nearly identically distributed with either definition of background. <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html">HTODemux function in Seurat</a> <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq">deMULTIplex function from Multiseq</a> <a href="https://github.com/statgen/demuxlet">demuxlet</a></p>
<p>Before using the demultiplexed negative drops in DSB as the <em>empty_drop_matrix</em> argument, negative drops (based on barcode) can be further QCd to remove drops with high RNA to eliminate potential cells from the ambient protein estimate. these functions should be run to demultiplex the <strong>raw</strong> output from cell ranger with slight subsetting (i.e. outs/raw_feature_bc_matrix - see workflow below) because more empty drops are included from which to estimate the negative distribution in the k-medoids function Seurat uses internally, this will also result in more background droplets defined as negative which will increase the confidence in the estimate of background used by dsb.</p>
<p><strong>see 10x data vignette discussed above and shown here <a href="https://github.com/MattPM/dsb/issues/9" class="uri">https://github.com/MattPM/dsb/issues/9</a> </strong> <strong>please see vignettes in the “articles” tab at <a href="https://mattpm.github.io/dsb/" class="uri">https://mattpm.github.io/dsb/</a> for a detailed workflow detailing these steps</strong></p>
</div>
<div id="simple-example-workflow-seurat-version-3-for-experiments-with-hashing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-example-workflow-seurat-version-3-for-experiments-with-hashing-data" class="anchor"></a>Simple example workflow (Seurat Version 3) for experiments with Hashing data</h2>
<div class="sourceCode" id="cb9"><pre class="r"><span class="co"># raw = path to cell ranger outs/raw_feature_bc_matrix </span>

<span class="co"># slightly subset negative drops to include all with 5 uniwue mRNA </span>
<span class="no">seurat_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="no">raw</span>, <span class="kw">min.genes</span> <span class="kw">=</span> <span class="fl">5</span>)

<span class="co"># demultiplex </span>
<span class="no">seurat_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/HTODemux.html">HTODemux</a></span>(<span class="no">seurat_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"HTO"</span>, <span class="kw">positive.quantile</span> <span class="kw">=</span> <span class="fl">0.99</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(<span class="no">seurat_object</span>) <span class="kw">=</span> <span class="st">"HTO_classification.global"</span>

<span class="co"># subset empty drop/background and cells </span>
<span class="no">neg_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seurat_object</span>, <span class="kw">idents</span> <span class="kw">=</span> <span class="st">"Negative"</span>)
<span class="no">singlet_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seurat_object</span>, <span class="kw">idents</span> <span class="kw">=</span> <span class="st">"Singlet"</span>)

<span class="co"># non sparse CITEseq data actually store better in a regular materix so the as.matrix() call is not memory intensive.</span>
<span class="no">neg_adt_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(<span class="no">neg_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"CITE"</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">'counts'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()
<span class="no">positive_adt_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(<span class="no">singlet_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"CITE"</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">'counts'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()

<span class="co"># normalize the data with dsb</span>
<span class="no">dsb_norm_prot</span> <span class="kw">=</span> <span class="fu"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">positive_adt_matrix</span>,
                                    <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">neg_adt_matrix</span>)

<span class="co"># now add the normalized dat back to the object (the singlets defined above as "object")</span>
<span class="no">singlet_object</span><span class="kw">[[</span><span class="st">"CITE"</span>]] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateAssayObject.html">CreateAssayObject</a></span>(<span class="kw">data</span> <span class="kw">=</span> <span class="no">dsb_norm_prot</span>)</pre></div>
</div>
</div>
<div id="faq" class="section level1">
<h1 class="hasAnchor">
<a href="#faq" class="anchor"></a>FAQ</h1>
<p><strong>I get the error “Error in quantile.default(x, seq(from = 0, to = 1, length = n)): missing values and NaN’s not allowed if ‘na.rm’ is FALSE” What should I do?</strong> - (see issue 6) this error occurs during denoising, (denoise = TRUE) when you have antibodies with 0 counts or close to 0 across <em>all cells</em>. To get rid of this error, check the distributions of the antibodies with e.g. <code><a href="https://rdrr.io/r/base/apply.html">apply(cells_protein_matrix, 1, quantile)</a></code> to find the protein(s) with basically no counts, then remove these from the empty drops and the cells. (see issue 5)</p>
<p><strong>I get a "problem too large or memory exhausted error when I try to convert to a regular R matrix</strong> - (see issue 10) CITE-seq protein counts don’t need a sparse representation-very likely this error is because there are too many negative droplets defined (i.e. over 1 million). You should be able to normalize datasets with 100,000+ cells and similar numbers of negative droplets (or less) on a normal 16GB laptop. By further narrowing in on the major background distribution, one should be able to convert the cells and background to a normal R matrix which should run successfully.</p>
<p><strong>I have 2 batches, should I combine them into a single batch or normalize each batc hseparately?</strong> - (See issue 12) How much batch variation there is depends on how much experiment-specific and expected biological variability there is between the batches. In the dataset used in the preprint, if we normalized with all background drops and cells in a single normalization, the resulting dsb normalized values were highly concordant with when we normalized each batch separately, this held true with either definition of background drops used (i.e. based on thresholding with the library size or based on hashing-see below). One could try both and see which mitigates a the batch variation the most. See issue 12 for example code to do this.<br><a href="https://mattpm.github.io/dsb"><img src="reference/figures/single_multi.png" height="300"></a></p>
<div id="niaid-statement" class="section level3">
<h3 class="hasAnchor">
<a href="#niaid-statement" class="anchor"></a>NIAID statement</h3>
<p>A review of this code has been conducted, no critical errors exist, and to the best of the authors knowledge, there are no problematic file paths, no local system configuration details, and no passwords or keys included in this code. This open source software comes as is with absolutely no warranty.</p>
<p>###Terms of Use By using this software, you agree this software is to be used for research purposes only. Any presentation of data analysis using dsb will acknowledge the software according to the guidelines below.</p>
<p>Citation: 1. Mulè MP, Martins AJ, Tsang JS. Normalizing and denoising protein expression data from droplet-based single cell profiling. bioRxiv. 2020;2020.02.24.963603.</p>
<p>Primary author(s): Matt Mulè Organizational contact information: General: john.tsang AT nih.gov, code: mulemp AT nih.gov<br>
Date of release: Oct 7 2020<br>
Version: 0.1.0 License details: see package Description: code to reproduce analysis of manuscript<br>
Usage instructions: Provided in this markdown<br>
Example(s) of usage: NA<br>
Proper attribution to others, when applicable: NA</p>
</div>
<div id="code-check" class="section level3">
<h3 class="hasAnchor">
<a href="#code-check" class="anchor"></a>code check</h3>
<div class="sourceCode" id="cb10"><pre class="r">
<span class="co"># checked repo for PII and searched for strings with paths </span>

<span class="co"># code check </span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">lintr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">here</span>)
<span class="no">fcn</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/here/man/here.html">here</a></span>(<span class="st">"R/"</span>), <span class="st">".r"</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>))
<span class="no">vignette</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span>(<span class="kw">path</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/here/man/here.html">here</a></span>(<span class="st">"vignettes/dsb_normalizing_CITEseq_data.R"</span>))
<span class="co"># code check </span>
<span class="no">scp</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="no">fcn</span>, <span class="no">vignette</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">as.list</a></span>()
<span class="no">lt</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>((<span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="no">scp</span>, <span class="kw pkg">lintr</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/lintr/man/lint_file.html">lint</a></span>)))

<span class="co"># ignoring 80 character line warning - this is not applicable to documentation can't break lines with roxygen syntax </span></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/MattPM/dsb/">https://​github.com/​MattPM/​dsb/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/MattPM/dsb/issues">https://​github.com/​MattPM/​dsb/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li> <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
