<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Normalize CITEseq Data • dsb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Normalize CITEseq Data">
<meta property="og:description" content="see biorxiv preprint DOI:Here we developed a method specifically for normalizing CITEseq data,accounting for protein-specific background noise caused by unbound antibody captured and sequenced in droplets as well as correcting for the technical component of variation in protein library size. We account for these sources of experimental noisewith the empty droplets containing background antibody reads and by defining a covariate of non-staining proteins and isotype control antibodies respectively.">
<meta property="og:image" content="https://mattpm.github.io/dsb/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">dsb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/dsb_normalizing_CITEseq_data.html">dsb_normalizing_CITEseq_data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/dsb">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="dsb" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#dsb" class="anchor"></a>dsb <a href="https://mattpm.github.io/dsb"><img src="reference/figures/logo.png" align="right" height="150"></a>
</h1></div>
<div id="an-r-package-for-normalizing-and-denoising-citeseq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#an-r-package-for-normalizing-and-denoising-citeseq-data" class="anchor"></a>An R package for normalizing and denoising CITEseq data</h2>
<!-- badges: start -->
<!-- badges: end -->
<p>This package was developed at <a href="https://www.niaid.nih.gov/research/john-tsang-phd">John Tsang’s Lab</a> by Matt Mulè and Andrew Martins and John Tsang. The package implements our normalization and denoising method for CITEseq data. Technical discussion of how the method works can be found in <a href="https://biorxiv.org">the biorxiv preprint</a> We utilized the dsb package to normalize CITEseq data reported in this paper <a href="https://"></a></p>
<p>In <a href="https://biorxiv.org">the biorxiv preprint</a>, comparing unstained control cells and empty droplets we found the major contribotor to background noise in CITEseq data is unbound antibody captured and sequenced in droplets. DSB corrects for this background by leveraging empty droplets which serve as a “built in” noise measurement in any droplet capture single cell platform (e.g. 10X, dropseq, indrop).</p>
<p>In addition we define a per-cell denoising covariate to account for the technical component of library size differences between cells which removes spurious cluster formation derived from globally dimcells clustering together.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>installation</h2>
<p>You can install the released version of dsb in your R session with the command below</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># this is analagous to install.packages("package), you need the package devtools to install a package from a github repository like this one. </span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(devtools)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">#&gt; Loading required package: devtools</span></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="co">#devtools::install_github(repo = 'MattPM/dsb')</span></a></code></pre></div>
</div>
<div id="quickstart" class="section level2">
<h2 class="hasAnchor">
<a href="#quickstart" class="anchor"></a>quickstart</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># load package and normalize the example raw data </span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dsb)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># normalize</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> cells_citeseq_mtx,</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">                                        <span class="dt">empty_drop_matrix =</span> empty_drop_citeseq_mtx)</a></code></pre></div>
</div>
<div id="the-full-version-recommended" class="section level2">
<h2 class="hasAnchor">
<a href="#the-full-version-recommended" class="anchor"></a>The full version (recommended)</h2>
<p>By default dsb defines the per cell denoising covariate by fitting a gaussian mixture model to the log + 10 counts of each cell and defining the noise ocvariates the mean. We reccomend including the counts from isotype controls in each cell in the denoising covariates.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co"># get the empty cells from demultiplexing with </span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co"># define a vector of the isotype controls in the data </span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">isotypes =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Mouse IgG2bkIsotype_PROT"</span>, <span class="st">"MouseIgG1kappaisotype_PROT"</span>,<span class="st">"MouseIgG2akappaisotype_PROT"</span>, <span class="st">"RatIgG2bkIsotype_PROT"</span>)</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> cells_citeseq_mtx,</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">                                        <span class="dt">empty_drop_matrix =</span> empty_drop_citeseq_mtx,</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">                                        <span class="dt">use.isotype.control =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb3-11" data-line-number="11">                                        <span class="dt">isotype.control.name.vec =</span> isotypes)</a></code></pre></div>
</div>
<div id="visualize-distributions-of-cd4-and-cd8" class="section level2">
<h2 class="hasAnchor">
<a href="#visualize-distributions-of-cd4-and-cd8" class="anchor"></a>visualize distributions of CD4 and CD8</h2>
<p>plot the DSB normalized CITEseq data.</p>
<p><strong>Note, there is NO jitter added to these points for visualization these are the unmodified normalized counts</strong></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># add a density gradient on the points () this is helpful when there are many thousands of cells )</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># this density function is from this blog post: https://slowkow.com/notes/ggplot2-color-by-density/</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">get_density =<span class="st"> </span><span class="cf">function</span>(x, y, ...) {</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">  dens &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/MASS/man/kde2d.html">kde2d</a></span>(x, y, ...)</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">  ix &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/findInterval.html">findInterval</a></span>(x, dens<span class="op">$</span>x)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">  iy &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/findInterval.html">findInterval</a></span>(y, dens<span class="op">$</span>y)</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">  ii &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(ix, iy)</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(dens<span class="op">$</span>z[ii])</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">}</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">data.plot =<span class="st"> </span>normalized_matrix <span class="op">%&gt;%</span><span class="st"> </span>t <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(CD4_PROT, CD8_PROT, CD27_PROT, CD19_PROT) </a>
<a class="sourceLine" id="cb4-15" data-line-number="15"></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"></a>
<a class="sourceLine" id="cb4-18" data-line-number="18">data.plot =<span class="st"> </span>data.plot <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">density =</span> <span class="kw">get_density</span>(data.plot<span class="op">$</span>CD4_PROT, data.plot<span class="op">$</span>CD8_PROT, <span class="dt">n =</span> <span class="dv">100</span>)) </a>
<a class="sourceLine" id="cb4-19" data-line-number="19">p1 =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(data.plot, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> CD8_PROT, <span class="dt">y =</span> CD4_PROT, <span class="dt">color =</span> density)) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">size =</span> <span class="fl">0.4</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"small example dataset"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="dt">xintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-22" data-line-number="22"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-23" data-line-number="23"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">face =</span> <span class="st">"bold"</span>,<span class="dt">size =</span> <span class="dv">12</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-24" data-line-number="24"><span class="st">  </span>viridis<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/viridis/man/scale_viridis.html">scale_color_viridis</a></span>(<span class="dt">option =</span> <span class="st">"B"</span>) <span class="op">+</span><span class="st">  </span></a>
<a class="sourceLine" id="cb4-25" data-line-number="25"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_identity.html">scale_shape_identity</a></span>() </a>
<a class="sourceLine" id="cb4-26" data-line-number="26"></a>
<a class="sourceLine" id="cb4-27" data-line-number="27"></a>
<a class="sourceLine" id="cb4-28" data-line-number="28">data.plot =<span class="st"> </span>data.plot <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">density =</span> <span class="kw">get_density</span>(data.plot<span class="op">$</span>CD19_PROT, data.plot<span class="op">$</span>CD27_PROT, <span class="dt">n =</span> <span class="dv">100</span>)) </a>
<a class="sourceLine" id="cb4-29" data-line-number="29">p2 =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(data.plot, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> CD19_PROT, <span class="dt">y =</span> CD27_PROT, <span class="dt">color =</span> density)) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-30" data-line-number="30"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">size =</span> <span class="fl">0.4</span>) <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-31" data-line-number="31"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="dt">xintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-32" data-line-number="32"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-33" data-line-number="33"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">face =</span> <span class="st">"bold"</span>,<span class="dt">size =</span> <span class="dv">12</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-34" data-line-number="34"><span class="st">  </span>viridis<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/viridis/man/scale_viridis.html">scale_color_viridis</a></span>(<span class="dt">option =</span> <span class="st">"B"</span>) <span class="op">+</span><span class="st">  </span></a>
<a class="sourceLine" id="cb4-35" data-line-number="35"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_identity.html">scale_shape_identity</a></span>() </a>
<a class="sourceLine" id="cb4-36" data-line-number="36"></a>
<a class="sourceLine" id="cb4-37" data-line-number="37">cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(p1,p2)</a></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-4-1.png" width="100%"></p>
</div>
<div id="how-do-i-get-the-empty-droplets" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-i-get-the-empty-droplets" class="anchor"></a>How do I get the empty droplets?</h2>
<p>There are a number of ways to get the empty drops. If you are using cell hashing, when you demultiplex the cells, you get a vector of empty or Negative droplets.</p>
<p>HTODemux function in Seurat: <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html" class="uri">https://satijalab.org/seurat/v3.1/hashing_vignette.html</a></p>
<p>deMULTIplex function from Multiseq (this is now also implemented in Seurat). <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq" class="uri">https://github.com/chris-mcginnis-ucsf/MULTI-seq</a></p>
<p>If you’re not multiplexing you can simply get a vector of negative droplets from the cells you would remove. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y" class="uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y</a></p>
</div>
<div id="simple-example-workflow-seurat-version-3" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-example-workflow-seurat-version-3" class="anchor"></a>Simple example workflow (Seurat Version 3)</h2>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co"># get the ADT counts using Seurat version 3 </span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">seurat_object =<span class="st"> </span><span class="kw">HTODemux</span>(seurat_object, <span class="dt">assay =</span> <span class="st">"HTO"</span>, <span class="dt">positive.quantile =</span> <span class="fl">0.99</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="kw">Idents</span>(seurat_object) =<span class="st"> "HTO_classification.global"</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5">neg_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(seurat_object, <span class="dt">idents =</span> <span class="st">"Negative"</span>)</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">singlet_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(seurat_object, <span class="dt">idents =</span> <span class="st">"Singlet"</span>)</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"></a>
<a class="sourceLine" id="cb5-9" data-line-number="9"><span class="co"># non sparse CITEseq data actually store better in a regular materix so the as.matrix() call is not memory intensive.</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10">neg_adt_matrix =<span class="st"> </span><span class="kw">GetAssayData</span>(neg_object, <span class="dt">assay =</span> <span class="st">"CITE"</span>, <span class="dt">slot =</span> <span class="st">'raw.data'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb5-11" data-line-number="11">positive_adt_matrix =<span class="st"> </span><span class="kw">GetAssayData</span>(singlet_object, <span class="dt">assay =</span> <span class="st">"CITE"</span>, <span class="dt">slot =</span> <span class="st">'raw.data'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb5-12" data-line-number="12"></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"></a>
<a class="sourceLine" id="cb5-14" data-line-number="14"><span class="co"># normalize the data with dsb</span></a>
<a class="sourceLine" id="cb5-15" data-line-number="15"><span class="co"># make sure you've run devtools::install_github(repo = 'MattPM/dsb')</span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> positive_adt_matrix,</a>
<a class="sourceLine" id="cb5-17" data-line-number="17">                                        <span class="dt">empty_drop_matrix =</span> neg_adt_matrix)</a>
<a class="sourceLine" id="cb5-18" data-line-number="18"></a>
<a class="sourceLine" id="cb5-19" data-line-number="19"></a>
<a class="sourceLine" id="cb5-20" data-line-number="20"><span class="co"># now add the normalized dat back to the object (the singlets defined above as "object")</span></a>
<a class="sourceLine" id="cb5-21" data-line-number="21">singlet_object =<span class="st"> </span><span class="kw">SetAssayData</span>(<span class="dt">object =</span> singlet_object, <span class="dt">slot =</span> <span class="st">"CITE"</span>, <span class="dt">new.data =</span> normalized_matrix)</a></code></pre></div>
<p>In reality you might want to confirm the cells called as “Negative” have low RNA / gene content to be certain there are no contaminating cells.</p>
<p>Also it is not necessary but we reccomend hash demultiplexing with the raw output from cellranger rather than the processed output (i.e. outs/raw_feature_bc_matrix) will have more empty droplets from which the HTODemux function will be able to estimate the negative distribution. This will also have the benefit of creating more droplets to use as built protein background controls in the DSB function.</p>
</div>
<div id="example-workflow-seurat-version-2" class="section level2">
<h2 class="hasAnchor">
<a href="#example-workflow-seurat-version-2" class="anchor"></a>example workflow Seurat version 2</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="co"># get the ADT counts using Seurat version 3 </span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">seurat_object =<span class="st"> </span><span class="kw">HTODemux</span>(seurat_object, <span class="dt">assay =</span> <span class="st">"HTO"</span>, <span class="dt">positive.quantile =</span> <span class="fl">0.99</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"></a>
<a class="sourceLine" id="cb6-5" data-line-number="5">neg =<span class="st"> </span>seurat_object <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="st">  </span><span class="kw">SetAllIdent</span>(<span class="dt">id =</span> <span class="st">"hto_classification_global"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="st">  </span><span class="kw">SubsetData</span>(<span class="dt">ident.use =</span> <span class="st">"Negative"</span>) </a>
<a class="sourceLine" id="cb6-8" data-line-number="8"></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">singlet =<span class="st"> </span>seurat_object <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-10" data-line-number="10"><span class="st">  </span><span class="kw">SetAllIdent</span>(<span class="dt">id =</span> <span class="st">"hto_classification_global"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="st">  </span><span class="kw">SubsetData</span>(<span class="dt">ident.use =</span> <span class="st">"Singlet"</span>) </a>
<a class="sourceLine" id="cb6-12" data-line-number="12"></a>
<a class="sourceLine" id="cb6-13" data-line-number="13"><span class="co"># get negative and positive ADT data </span></a>
<a class="sourceLine" id="cb6-14" data-line-number="14">neg_adt_matrix =<span class="st"> </span>neg<span class="op">@</span>assay<span class="op">$</span>CITE<span class="op">@</span>raw.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb6-15" data-line-number="15">pos_adt_matrix =<span class="st"> </span>singlet<span class="op">@</span>assay<span class="op">$</span>CITE<span class="op">@</span>raw.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb6-16" data-line-number="16"></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"></a>
<a class="sourceLine" id="cb6-18" data-line-number="18"><span class="co"># normalize the data with dsb</span></a>
<a class="sourceLine" id="cb6-19" data-line-number="19"><span class="co"># make sure you've run devtools::install_github(repo = 'MattPM/dsb')</span></a>
<a class="sourceLine" id="cb6-20" data-line-number="20">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> pos_adt_matrix,</a>
<a class="sourceLine" id="cb6-21" data-line-number="21">                                        <span class="dt">empty_drop_matrix =</span> neg_adt_matrix)</a>
<a class="sourceLine" id="cb6-22" data-line-number="22"></a>
<a class="sourceLine" id="cb6-23" data-line-number="23"></a>
<a class="sourceLine" id="cb6-24" data-line-number="24"><span class="co"># add the assay to the Seurat object </span></a>
<a class="sourceLine" id="cb6-25" data-line-number="25">singlet =<span class="st"> </span><span class="kw">SetAssayData</span>(<span class="dt">object =</span> singlet, <span class="dt">slot =</span> <span class="st">"CITE"</span>, <span class="dt">new.data =</span> normalized_matrix)</a></code></pre></div>
</div>
<div id="how-to-get-empty-drops-without-cell-hashing-or-sample-demultiplexing" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-get-empty-drops-without-cell-hashing-or-sample-demultiplexing" class="anchor"></a>How to get empty drops without cell hashing or sample demultiplexing</h2>
<p>you can simply get a vector of negative droplets from the cells you would remove. There robust ways to estimate which cells are empty droplets: <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y" class="uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y</a></p>
<p>Below is a quick method to get outlier empty droplets assuming seurat_object is a object with most cells (i.e. any cell expressing at least a gene).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># get the nUMI from a seurat version 3 object </span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">umi =<span class="st"> </span>seurat_object<span class="op">$</span>nUMI</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="co">#  Get the nUMI from a Seurat version 2 objec </span></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">umi =<span class="st"> </span>seurat_object<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="st">"nUMI"</span>)</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">mu_umi =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(umi)</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">sd_umi =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(umi)</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"></a>
<a class="sourceLine" id="cb7-9" data-line-number="9"><span class="co"># calculate a threshold for calling a cell negative </span></a>
<a class="sourceLine" id="cb7-10" data-line-number="10">sub_threshold =<span class="st"> </span>mu_umi <span class="op">-</span><span class="st"> </span>(<span class="dv">5</span><span class="op">*</span>sd_umi)</a>
<a class="sourceLine" id="cb7-11" data-line-number="11"></a>
<a class="sourceLine" id="cb7-12" data-line-number="12"></a>
<a class="sourceLine" id="cb7-13" data-line-number="13"></a>
<a class="sourceLine" id="cb7-14" data-line-number="14"><span class="kw">Idents</span>(seurat_object) =<span class="st"> "nUMI"</span></a>
<a class="sourceLine" id="cb7-15" data-line-number="15"></a>
<a class="sourceLine" id="cb7-16" data-line-number="16"></a>
<a class="sourceLine" id="cb7-17" data-line-number="17"><span class="co"># gdefine the negative cell object</span></a>
<a class="sourceLine" id="cb7-18" data-line-number="18">neg =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(seurat_object, <span class="dt">accept.high =</span> sub_threshold)</a></code></pre></div>
<p>This negative cell object can be used to define the negative background following the examples above.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/MattPM/dsb">https://​github.com/​MattPM/​dsb</a>
</li>
<li>Report a bug at <br><a href="https://github.com/MattPM/dsb/issues">https://​github.com/​MattPM/​dsb/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li> <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
