<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Normalize CITEseq Data • dsb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Normalize CITEseq Data">
<meta property="og:description" content="see biorxiv preprint DOI:Here we developed a method specifically for normalizing CITEseq data,accounting for protein-specific background noise caused by unbound antibody captured and sequenced in droplets as well as correcting for the technical component of variation in protein library size. We account for these sources of experimental noisewith the empty droplets containing background antibody reads and by defining a covariate of non-staining proteins and isotype control antibodies respectively.">
<meta property="og:image" content="https://mattpm.github.io/dsb/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">dsb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/dsb_normalizing_CITEseq_data.html">dsb_normalizing_CITEseq_data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/dsb">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="dsb" class="section level2">
<h2 class="hasAnchor">
<a href="#dsb" class="anchor"></a>dsb <a href="https://mattpm.github.io/dsb"><img src="reference/figures/logo.png" align="right" height="150"></a>
</h2>
</div>
<div id="an-r-package-for-normalizing-and-denoising-citeseq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#an-r-package-for-normalizing-and-denoising-citeseq-data" class="anchor"></a>An R package for normalizing and denoising CITEseq data</h2>
<!-- badges: start -->
<!-- [![Travis build status](https://travis-ci.org/MattPM/dsb.svg?branch=master)](https://travis-ci.org/MattPM/dsb) -->
<!-- badges: end -->
<p><strong>check out more vignettes on the <a href="https://mattpm.github.io/dsb/">dsb package site</a></strong></p>
<p><strong>DSB was recently used in <a href="https://www.biorxiv.org/content/10.1101/2020.06.15.153080v1">this informative preprint on optomizing CITE-seq experiments</a></strong></p>
<p>dsb is a lightweight R package and is the first dedicated method for normalizing and denoising protein expression data from CITE-seq experiments. The method was developed in <a href="https://www.niaid.nih.gov/research/john-tsang-phd">John Tsang’s Lab</a> by Matt Mulè, Andrew Martins and John Tsang. Details of the method can be found in <a href="https://www.biorxiv.org/content/10.1101/2020.02.24.963603v1.full.pdf">the biorxiv preprint</a>. We utilized the dsb package to normalize CITEseq data reported in <a href="https://doi.org/10.1038/s41591-020-0769-8">our Nature Medicine paper on baseline immune states</a>.</p>
<p>As described in <a href="https://www.biorxiv.org/content/10.1101/2020.02.24.963603v1.full.pdf">preprint</a> in experiments comparing unstained control cells and empty droplets we found a major contributor to background noise in CITE-seq protein data is unbound antibodies encapsulated and sequenced in droplets. DSB corrects for this background by leveraging empty droplets, which serve as a “built in” noise measurement in all droplet capture single cell experiments. In an optional recommended step we define and regress out for each cell the primary latent noise component of that cell’s protein library to account for several potential sources of technical differences among single cells – see our preprint for details.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># this is analagous to install.packages("package), you need the package devtools to install a package from a github repository like this one. </span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(devtools); devtools<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="dt">repo =</span> <span class="st">'MattPM/dsb'</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dsb)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="co"># quick example - RUN dsb on package example data </span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6">norm_mtx =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> cells_citeseq_mtx, <span class="dt">empty_drop_matrix =</span> empty_drop_citeseq_mtx)</a></code></pre></div>
</div>
<div id="quick-start-using-raw-10x-cellranger-output" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#quick-start-using-raw-10x-cellranger-output" class="anchor"></a>Quick start using <em>RAW</em> 10X cellranger output</h1></div>
<p>load RAW (not filtered!) <strong>feature / cellmatrix raw</strong> public 10X CITE-seq data from here: <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_protein_v3" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_protein_v3</a></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dsb)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Seurat) <span class="co"># version 3 in example provided belo</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(tidyverse)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(magrittr)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">path_to_data =<span class="st"> "data/10x_data/10x_pbmc5k_V3/raw_feature_bc_matrix/"</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">raw =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="dt">data.dir =</span> path_to_data)</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="co"># create object with Minimal filtering (retain drops with 5 unique mRNAs detected)</span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9">s1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="dt">counts =</span> raw<span class="op">$</span><span class="st">`</span><span class="dt">Gene Expression</span><span class="st">`</span>,  <span class="dt">min.cells =</span> <span class="dv">10</span>, <span class="dt">min.features =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb2-10" data-line-number="10"></a>
<a class="sourceLine" id="cb2-11" data-line-number="11"><span class="co"># add some metadata </span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12">s1<span class="op">$</span>log10umi =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(s1<span class="op">$</span>nCount_RNA  <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) </a>
<a class="sourceLine" id="cb2-13" data-line-number="13">s1<span class="op">$</span>bc =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(s1<span class="op">@</span>meta.data)</a>
<a class="sourceLine" id="cb2-14" data-line-number="14"></a>
<a class="sourceLine" id="cb2-15" data-line-number="15"><span class="co"># define negative and positive drops based on an mRNA threshold to define neg and positive cells (see details below) </span></a>
<a class="sourceLine" id="cb2-16" data-line-number="16"><span class="kw"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(s1<span class="op">$</span>nCount_RNA  <span class="op">+</span><span class="st"> </span><span class="dv">1</span>), <span class="dt">breaks =</span> <span class="dv">1000</span>)</a>
<a class="sourceLine" id="cb2-17" data-line-number="17">neg_drops =<span class="st"> </span>s1<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(log10umi <span class="op">&gt;</span><span class="st"> </span><span class="fl">1.5</span> <span class="op">&amp;</span><span class="st"> </span>log10umi <span class="op">&lt;</span><span class="st"> </span><span class="fl">2.79</span>) <span class="op">%$%</span><span class="st"> </span>bc</a>
<a class="sourceLine" id="cb2-18" data-line-number="18">positive_cells =<span class="st"> </span>s1<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(log10umi <span class="op">&gt;</span><span class="st"> </span><span class="fl">2.8</span> <span class="op">&amp;</span><span class="st"> </span>log10umi <span class="op">&lt;</span><span class="st"> </span><span class="fl">4.4</span>) <span class="op">%$%</span><span class="st"> </span>bc</a>
<a class="sourceLine" id="cb2-19" data-line-number="19">  </a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="co"># Subset protein data to create a standard R matrix of protein counts for negative droplets and cells </span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21">neg_prot =<span class="st"> </span>raw<span class="op">$</span><span class="st">`</span><span class="dt">Antibody Capture</span><span class="st">`</span>[ ,  neg_drops ] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb2-22" data-line-number="22">positive_prot =<span class="st"> </span>raw<span class="op">$</span><span class="st">`</span><span class="dt">Antibody Capture</span><span class="st">`</span>[ , positive_cells] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb2-23" data-line-number="23"></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co"># run DSB normalization</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25">isotypes =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(positive_prot)[<span class="dv">30</span><span class="op">:</span><span class="dv">32</span>]</a>
<a class="sourceLine" id="cb2-26" data-line-number="26">mtx =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> positive_prot,</a>
<a class="sourceLine" id="cb2-27" data-line-number="27">                           <span class="dt">empty_drop_matrix =</span> neg_prot,</a>
<a class="sourceLine" id="cb2-28" data-line-number="28">                           <span class="dt">denoise.counts =</span> <span class="ot">TRUE</span>, <span class="dt">use.isotype.control =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb2-29" data-line-number="29">                           <span class="dt">isotype.control.name.vec =</span> isotypes)</a>
<a class="sourceLine" id="cb2-30" data-line-number="30"></a>
<a class="sourceLine" id="cb2-31" data-line-number="31"><span class="co">## subset the raw data to only iinclude cell containing drops and add protein data </span></a>
<a class="sourceLine" id="cb2-32" data-line-number="32">s =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(s1, <span class="dt">cells =</span> <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(mtx))</a>
<a class="sourceLine" id="cb2-33" data-line-number="33">s[[<span class="st">"CITE"</span>]] =<span class="st"> </span><span class="kw">CreateAssayObject</span>(<span class="dt">counts =</span> positive_prot)</a>
<a class="sourceLine" id="cb2-34" data-line-number="34">s =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SetAssayData.html">SetAssayData</a></span>(s,<span class="dt">assay =</span> <span class="st">"CITE"</span>, <span class="dt">slot =</span> <span class="st">"data"</span>,<span class="dt">new.data =</span> mtx)</a>
<a class="sourceLine" id="cb2-35" data-line-number="35"></a>
<a class="sourceLine" id="cb2-36" data-line-number="36"><span class="co"># Done! you can also save the resulting normalized matrix for integration with scanpy etc</span></a>
<a class="sourceLine" id="cb2-37" data-line-number="37"><span class="kw">write_delim</span>(mtx, <span class="dt">path =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(path_to_data, <span class="st">"dsb_norm_matrix.txt"</span>),<span class="dt">delim =</span> <span class="st">"</span><span class="ch">\t</span><span class="st">"</span> )</a></code></pre></div>
</div>
<div id="recommended-next-steps-in-citeseq-workflow--protein-based-clustering-and-protein-based-based-cluster-annotation" class="section level1">
<h1 class="hasAnchor">
<a href="#recommended-next-steps-in-citeseq-workflow--protein-based-clustering-and-protein-based-based-cluster-annotation" class="anchor"></a>Recommended next steps in CITEseq workflow. Protein based clustering and protein based based cluster annotation</h1>
<p>This is the same process followed in our paper <a href="https://www.nature.com/articles/s41591-020-0769-8" class="uri">https://www.nature.com/articles/s41591-020-0769-8</a></p>
<p>DSB normalized vlaues provide a straightforward comparable value for each protein in each cluster. They are the denoised log number of standard deviations from the background.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="co"># Get dsb normalized protein data without isotype controls for clustering</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">s_dsb =<span class="st"> </span>s<span class="op">@</span>assays<span class="op">$</span>CITE<span class="op">@</span>data</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">s_dsb =<span class="st"> </span>s_dsb[<span class="dv">1</span><span class="op">:</span><span class="dv">29</span>, ]</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co"># defint euclidean distance matrix and cluster </span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7">p_dist =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/dist.html">dist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(s_dsb))</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">p_dist =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(p_dist)</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">s[[<span class="st">"p_dist"</span>]] &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(p_dist)<span class="op">$</span>snn</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">s =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/FindClusters.html">FindClusters</a></span>(s, <span class="dt">resolution =</span> <span class="fl">0.6</span>, <span class="dt">graph.name =</span> <span class="st">"p_dist"</span>)</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="co"># Plot clusters by average protein expression for annotation </span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"></a>
<a class="sourceLine" id="cb3-14" data-line-number="14">adt_data =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/t.html">t</a></span>(s<span class="op">@</span>assays<span class="op">$</span>CITE<span class="op">@</span>data)), s<span class="op">@</span>meta.data)</a>
<a class="sourceLine" id="cb3-15" data-line-number="15">prots =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(s<span class="op">@</span>assays<span class="op">$</span>CITE<span class="op">@</span>data)</a>
<a class="sourceLine" id="cb3-16" data-line-number="16">adt_plot =<span class="st"> </span>adt_data <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"><span class="st">    </span><span class="kw">group_by</span>(seurat_clusters) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="st">    </span><span class="kw">summarize_at</span>(<span class="dt">.vars =</span> prots, <span class="dt">.funs =</span> mean) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="st">    </span><span class="kw">column_to_rownames</span>(<span class="st">"seurat_clusters"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-20" data-line-number="20"><span class="st">    </span>t <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="st">    </span>as.data.frame</a>
<a class="sourceLine" id="cb3-22" data-line-number="22"></a>
<a class="sourceLine" id="cb3-23" data-line-number="23">pheatmap<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html">pheatmap</a></span>(adt_plot, <span class="dt">color =</span> viridis<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/viridis/man/reexports.html">viridis</a></span>(<span class="dv">25</span>, <span class="dt">option =</span> <span class="st">"B"</span>), <span class="dt">fontsize_row =</span> <span class="dv">8</span>)</a></code></pre></div>
<p><img src="images/cluster_average_dsb.png"></p>
</div>
<div id="more-information-how-were-background-drops-defined-in-the-quick-example-above" class="section level1">
<h1 class="hasAnchor">
<a href="#more-information-how-were-background-drops-defined-in-the-quick-example-above" class="anchor"></a>More information: How were background drops defined in the quick example above?</h1>
<p>First apply some minimal (RNA based in this case) filtering to retain noise / empty droplets for DSB. Below the cells that should be used for background in this experiment are shown in a histogram of the droplets passing a minimal filtering step.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">path_to_data =<span class="st"> "data/10x_data/10x_pbmc5k_V3/raw_feature_bc_matrix/"</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">raw =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="dt">data.dir =</span> path_to_data)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="co"># create object with Minimal filtering (retain drops with 5 unique mRNAs detected)</span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">s1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="dt">counts =</span> raw<span class="op">$</span><span class="st">`</span><span class="dt">Gene Expression</span><span class="st">`</span>,  <span class="dt">min.cells =</span> <span class="dv">10</span>, <span class="dt">min.features =</span> <span class="dv">5</span>)</a>
<a class="sourceLine" id="cb4-7" data-line-number="7"></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co"># define number of total drops (&gt;130K) after minimal filtering</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9">ndrop =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(s1<span class="op">@</span>meta.data)[<span class="dv">1</span>]</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co"># Plot</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12">hist_attr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(  <span class="kw">theme_bw</span>() , <span class="kw">theme</span>(<span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>)) , <span class="kw">geom_density</span>(<span class="dt">fill =</span> <span class="st">"#3e8ede"</span>) )</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">p1 =<span class="st"> </span><span class="kw">ggplot</span>(s1<span class="op">@</span>meta.data, <span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(nCount_RNA <span class="op">+</span><span class="st"> </span><span class="dv">1</span> ) )) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="st">  </span>hist_attr <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>( <span class="st">" raw_feature_bc_matrix: "</span>, ndrop, <span class="st">" droplets"</span>)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2.8</span>, <span class="fl">1.4</span> ),   <span class="dt">linetype =</span> <span class="st">"dashed"</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="st">  </span><span class="kw">annotate</span>(<span class="st">"text"</span>, <span class="dt">x =</span> <span class="dv">1</span>, <span class="dt">y=</span><span class="fl">1.5</span>, <span class="dt">label =</span> <span class="st">" region 1: </span><span class="ch">\n</span><span class="st"> void of data "</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="st">  </span><span class="kw">annotate</span>(<span class="st">"text"</span>, <span class="dt">x =</span> <span class="dv">2</span>, <span class="dt">y=</span><span class="dv">2</span>, <span class="dt">label =</span> <span class="st">" region 2: </span><span class="ch">\n</span><span class="st"> background drops </span><span class="ch">\n</span><span class="st"> define 'empty_drop_matrix' </span><span class="ch">\n</span><span class="st"> with these drops "</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="st">  </span><span class="kw">annotate</span>(<span class="st">"text"</span>, <span class="dt">x =</span> <span class="dv">4</span>, <span class="dt">y=</span><span class="dv">2</span>, <span class="dt">label =</span> <span class="st">" region 3: </span><span class="ch">\n</span><span class="st"> cell containing droplets </span><span class="ch">\n</span><span class="st"> zomed in on next plot"</span>) </a>
<a class="sourceLine" id="cb4-20" data-line-number="20"></a>
<a class="sourceLine" id="cb4-21" data-line-number="21"></a>
<a class="sourceLine" id="cb4-22" data-line-number="22">p2 =<span class="st"> </span><span class="kw">ggplot</span>(s1<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(nCount_RNA <span class="op">+</span><span class="st"> </span><span class="dv">1</span>) <span class="op">&gt;</span><span class="st"> </span><span class="fl">2.8</span>), <span class="kw">aes</span>(<span class="dt">x =</span> <span class="kw"><a href="https://rdrr.io/r/base/Log.html">log10</a></span>(nCount_RNA <span class="op">+</span><span class="st"> </span><span class="dv">1</span> ) )) <span class="op">+</span></a>
<a class="sourceLine" id="cb4-23" data-line-number="23"><span class="st">  </span>hist_attr <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-24" data-line-number="24"><span class="st">  </span><span class="kw">ggtitle</span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">" drops containing cells "</span>))  </a>
<a class="sourceLine" id="cb4-25" data-line-number="25">p3 =<span class="st"> </span>cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>( p1 , p2 ) </a>
<a class="sourceLine" id="cb4-26" data-line-number="26">p3</a></code></pre></div>
<p><img src="images/drop_distribition.png"></p>
<p>If there were no isotype controls in the example above, the call would have been:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">mtx =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> pos_prot,</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">                          <span class="dt">empty_drop_matrix =</span> neg_prot,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">                          <span class="dt">denoise.counts =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">                          <span class="dt">use.isotype.control =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<div id="quickstart-2-using-example-data-removing-background-as-captured-by-data-from-empty-droplets" class="section level2">
<h2 class="hasAnchor">
<a href="#quickstart-2-using-example-data-removing-background-as-captured-by-data-from-empty-droplets" class="anchor"></a>Quickstart 2 using example data; removing background as captured by data from empty droplets</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># load package and normalize the example raw data </span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dsb)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co"># normalize</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> cells_citeseq_mtx,</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">                                        <span class="dt">empty_drop_matrix =</span> empty_drop_citeseq_mtx)</a></code></pre></div>
</div>
<div id="quickstart-3-using-example-data-removing-background-and-correcting-for-per-cell-technical-factor-as-a-covariate" class="section level2">
<h2 class="hasAnchor">
<a href="#quickstart-3-using-example-data-removing-background-and-correcting-for-per-cell-technical-factor-as-a-covariate" class="anchor"></a>Quickstart 3 using example data; – removing background and correcting for per-cell technical factor as a covariate</h2>
<p>By default, dsb defines the per-cell technical covariate by fitting a two-component gaussian mixture model to the log + 10 counts (of all proteins) within each cell and defining the covariate as the mean of the “negative” component. We recommend also to use the counts from the isotype controls in each cell to compute the denoising covariate (defined as the first principal component of the isotype control counts and the “negative” count inferred by the mixture model above.)</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"></a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="co"># define a vector of the isotype controls in the data </span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3">isotypes =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Mouse IgG2bkIsotype_PROT"</span>, <span class="st">"MouseIgG1kappaisotype_PROT"</span>,</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">             <span class="st">"MouseIgG2akappaisotype_PROT"</span>, <span class="st">"RatIgG2bkIsotype_PROT"</span>)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5"></a>
<a class="sourceLine" id="cb7-6" data-line-number="6">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> cells_citeseq_mtx,</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">                                        <span class="dt">empty_drop_matrix =</span> empty_drop_citeseq_mtx,</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">                                        <span class="dt">use.isotype.control =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">                                        <span class="dt">isotype.control.name.vec =</span> isotypes)</a></code></pre></div>
</div>
<div id="visualization-on-example-data-distributions-of-cd4-and-cd8-dsb-normalized-citeseq-data-" class="section level2">
<h2 class="hasAnchor">
<a href="#visualization-on-example-data-distributions-of-cd4-and-cd8-dsb-normalized-citeseq-data-" class="anchor"></a>Visualization on example data: distributions of CD4 and CD8 DSB normalized CITEseq data.</h2>
<p><strong>Note, there is NO jitter added to these points for visualization; these are the unmodified normalized counts</strong></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># add a density gradient on the points () this is helpful when there are many thousands of cells )</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co"># this density function is from this blog post: https://slowkow.com/notes/ggplot2-color-by-density/</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">get_density =<span class="st"> </span><span class="cf">function</span>(x, y, ...) {</a>
<a class="sourceLine" id="cb8-4" data-line-number="4">  dens &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/MASS/man/kde2d.html">kde2d</a></span>(x, y, ...)</a>
<a class="sourceLine" id="cb8-5" data-line-number="5">  ix &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/findInterval.html">findInterval</a></span>(x, dens<span class="op">$</span>x)</a>
<a class="sourceLine" id="cb8-6" data-line-number="6">  iy &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/findInterval.html">findInterval</a></span>(y, dens<span class="op">$</span>y)</a>
<a class="sourceLine" id="cb8-7" data-line-number="7">  ii &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(ix, iy)</a>
<a class="sourceLine" id="cb8-8" data-line-number="8">  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(dens<span class="op">$</span>z[ii])</a>
<a class="sourceLine" id="cb8-9" data-line-number="9">}</a>
<a class="sourceLine" id="cb8-10" data-line-number="10"></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)</a>
<a class="sourceLine" id="cb8-12" data-line-number="12">data.plot =<span class="st"> </span>normalized_matrix <span class="op">%&gt;%</span><span class="st"> </span>t <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb8-13" data-line-number="13"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-14" data-line-number="14"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(CD4_PROT, CD8_PROT, CD27_PROT, CD19_PROT) </a>
<a class="sourceLine" id="cb8-15" data-line-number="15"></a>
<a class="sourceLine" id="cb8-16" data-line-number="16">density_attr =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(</a>
<a class="sourceLine" id="cb8-17" data-line-number="17">  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span>(<span class="dt">xintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>, <span class="dt">linetype =</span> <span class="dv">2</span>), </a>
<a class="sourceLine" id="cb8-18" data-line-number="18">  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">color =</span> <span class="st">"red"</span>, <span class="dt">linetype =</span> <span class="dv">2</span>), </a>
<a class="sourceLine" id="cb8-19" data-line-number="19">  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span>(<span class="dt">axis.text =</span> <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span>(<span class="dt">face =</span> <span class="st">"bold"</span>,<span class="dt">size =</span> <span class="dv">12</span>)) , </a>
<a class="sourceLine" id="cb8-20" data-line-number="20">  viridis<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/viridis/man/scale_viridis.html">scale_color_viridis</a></span>(<span class="dt">option =</span> <span class="st">"B"</span>), </a>
<a class="sourceLine" id="cb8-21" data-line-number="21">  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/scale_identity.html">scale_shape_identity</a></span>(), </a>
<a class="sourceLine" id="cb8-22" data-line-number="22">  <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>() </a>
<a class="sourceLine" id="cb8-23" data-line-number="23">)</a>
<a class="sourceLine" id="cb8-24" data-line-number="24"></a>
<a class="sourceLine" id="cb8-25" data-line-number="25"></a>
<a class="sourceLine" id="cb8-26" data-line-number="26">data.plot =<span class="st"> </span>data.plot <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">density =</span> <span class="kw">get_density</span>(data.plot<span class="op">$</span>CD4_PROT, data.plot<span class="op">$</span>CD8_PROT, <span class="dt">n =</span> <span class="dv">100</span>)) </a>
<a class="sourceLine" id="cb8-27" data-line-number="27">p1 =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(data.plot, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> CD8_PROT, <span class="dt">y =</span> CD4_PROT, <span class="dt">color =</span> density)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-28" data-line-number="28"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">size =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>density_attr <span class="op">+</span><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"small example dataset"</span>)</a>
<a class="sourceLine" id="cb8-29" data-line-number="29"></a>
<a class="sourceLine" id="cb8-30" data-line-number="30">data.plot =<span class="st"> </span>data.plot <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>(<span class="dt">density =</span> <span class="kw">get_density</span>(data.plot<span class="op">$</span>CD19_PROT, data.plot<span class="op">$</span>CD27_PROT, <span class="dt">n =</span> <span class="dv">100</span>)) </a>
<a class="sourceLine" id="cb8-31" data-line-number="31">p2 =<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(data.plot, <span class="kw"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="dt">x =</span> CD19_PROT, <span class="dt">y =</span> CD27_PROT, <span class="dt">color =</span> density)) <span class="op">+</span></a>
<a class="sourceLine" id="cb8-32" data-line-number="32"><span class="st">  </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="dt">size =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span>density_attr <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span>(<span class="st">"small example dataset"</span>)</a>
<a class="sourceLine" id="cb8-33" data-line-number="33"></a>
<a class="sourceLine" id="cb8-34" data-line-number="34">cowplot<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(p1,p2)</a></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-8-1.png" width="100%"></p>
</div>
<div id="how-do-i-get-the-empty-droplets" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-i-get-the-empty-droplets" class="anchor"></a>How do I get the empty droplets?</h2>
<p>If you don’t have hashing data, you can define the negative drops as shown above in the vignette using 10X data. If you have hashing data, demultiplexing functions define a “negative” cell population which can be used to define background.</p>
<p>HTODemux function in Seurat: <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html" class="uri">https://satijalab.org/seurat/v3.1/hashing_vignette.html</a></p>
<p>deMULTIplex function from Multiseq (this is now also implemented in Seurat). <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq" class="uri">https://github.com/chris-mcginnis-ucsf/MULTI-seq</a></p>
<p>In practice, you would want to confirm that the cells called as “negative” indeed have low RNA / gene content to be certain that there are no contaminating cells. Also, we recommend hash demultiplexing with the <em>raw</em> output from cellranger rather than the processed output (i.e. outs/raw_feature_bc_matrix). This output contains all barcodes and will have more empty droplets from which the HTODemux function will be able to estimate the negative distribution. This will also have the benefit of creating more empty droplets to use as built-in protein background controls in the DSB function.</p>
<p><strong>see 10x data vignette discussed above and shown here <a href="https://github.com/MattPM/dsb/issues/9" class="uri">https://github.com/MattPM/dsb/issues/9</a> </strong> <strong>please see vignettes in the “articles” tab at <a href="https://mattpm.github.io/dsb/" class="uri">https://mattpm.github.io/dsb/</a> for a detailed workflow detailing these steps</strong></p>
</div>
<div id="simple-example-workflow-seurat-version-3-for-experiments-with-hashing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-example-workflow-seurat-version-3-for-experiments-with-hashing-data" class="anchor"></a>Simple example workflow (Seurat Version 3) for experiments with Hashing data</h2>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="co"># get the ADT counts using Seurat version 3 </span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3">seurat_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/HTODemux.html">HTODemux</a></span>(seurat_object, <span class="dt">assay =</span> <span class="st">"HTO"</span>, <span class="dt">positive.quantile =</span> <span class="fl">0.99</span>)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"><span class="kw">Idents</span>(seurat_object) =<span class="st"> "HTO_classification.global"</span></a>
<a class="sourceLine" id="cb9-5" data-line-number="5">neg_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(seurat_object, <span class="dt">idents =</span> <span class="st">"Negative"</span>)</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">singlet_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/subset.html">subset</a></span>(seurat_object, <span class="dt">idents =</span> <span class="st">"Singlet"</span>)</a>
<a class="sourceLine" id="cb9-7" data-line-number="7"></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="co"># non sparse CITEseq data actually store better in a regular materix so the as.matrix() call is not memory intensive.</span></a>
<a class="sourceLine" id="cb9-10" data-line-number="10">neg_adt_matrix =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(neg_object, <span class="dt">assay =</span> <span class="st">"CITE"</span>, <span class="dt">slot =</span> <span class="st">'counts'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb9-11" data-line-number="11">positive_adt_matrix =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(singlet_object, <span class="dt">assay =</span> <span class="st">"CITE"</span>, <span class="dt">slot =</span> <span class="st">'counts'</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb9-12" data-line-number="12"></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"></a>
<a class="sourceLine" id="cb9-14" data-line-number="14"><span class="co"># normalize the data with dsb</span></a>
<a class="sourceLine" id="cb9-15" data-line-number="15"><span class="co"># make sure you've run devtools::install_github(repo = 'MattPM/dsb')</span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> positive_adt_matrix,</a>
<a class="sourceLine" id="cb9-17" data-line-number="17">                                        <span class="dt">empty_drop_matrix =</span> neg_adt_matrix)</a>
<a class="sourceLine" id="cb9-18" data-line-number="18"></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"></a>
<a class="sourceLine" id="cb9-20" data-line-number="20"><span class="co"># now add the normalized dat back to the object (the singlets defined above as "object")</span></a>
<a class="sourceLine" id="cb9-21" data-line-number="21">singlet_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SetAssayData.html">SetAssayData</a></span>(<span class="dt">object =</span> singlet_object, <span class="dt">slot =</span> <span class="st">"CITE"</span>, <span class="dt">new.data =</span> normalized_matrix)</a></code></pre></div>
</div>
<div id="simple-example-workflow-seurat-version-2-for-experiments-with-hashing-data" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-example-workflow-seurat-version-2-for-experiments-with-hashing-data" class="anchor"></a>Simple example workflow Seurat version 2 for experiments with hashing data</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="co"># get the ADT counts using Seurat version 3 </span></a>
<a class="sourceLine" id="cb10-3" data-line-number="3">seurat_object =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/HTODemux.html">HTODemux</a></span>(seurat_object, <span class="dt">assay =</span> <span class="st">"HTO"</span>, <span class="dt">positive.quantile =</span> <span class="fl">0.99</span>)</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"></a>
<a class="sourceLine" id="cb10-5" data-line-number="5">neg =<span class="st"> </span>seurat_object <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SetAllIdent.html">SetAllIdent</a></span>(<span class="dt">id =</span> <span class="st">"hto_classification_global"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SubsetData.html">SubsetData</a></span>(<span class="dt">ident.use =</span> <span class="st">"Negative"</span>) </a>
<a class="sourceLine" id="cb10-8" data-line-number="8"></a>
<a class="sourceLine" id="cb10-9" data-line-number="9">singlet =<span class="st"> </span>seurat_object <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb10-10" data-line-number="10"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SetAllIdent.html">SetAllIdent</a></span>(<span class="dt">id =</span> <span class="st">"hto_classification_global"</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb10-11" data-line-number="11"><span class="st">  </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SubsetData.html">SubsetData</a></span>(<span class="dt">ident.use =</span> <span class="st">"Singlet"</span>) </a>
<a class="sourceLine" id="cb10-12" data-line-number="12"></a>
<a class="sourceLine" id="cb10-13" data-line-number="13"><span class="co"># get negative and positive ADT data </span></a>
<a class="sourceLine" id="cb10-14" data-line-number="14">neg_adt_matrix =<span class="st"> </span>neg<span class="op">@</span>assay<span class="op">$</span>CITE<span class="op">@</span>raw.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb10-15" data-line-number="15">pos_adt_matrix =<span class="st"> </span>singlet<span class="op">@</span>assay<span class="op">$</span>CITE<span class="op">@</span>raw.data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()</a>
<a class="sourceLine" id="cb10-16" data-line-number="16"></a>
<a class="sourceLine" id="cb10-17" data-line-number="17"></a>
<a class="sourceLine" id="cb10-18" data-line-number="18"><span class="co"># normalize the data with dsb</span></a>
<a class="sourceLine" id="cb10-19" data-line-number="19"><span class="co"># make sure you've run devtools::install_github(repo = 'MattPM/dsb')</span></a>
<a class="sourceLine" id="cb10-20" data-line-number="20">normalized_matrix =<span class="st"> </span><span class="kw"><a href="reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="dt">cell_protein_matrix =</span> pos_adt_matrix,</a>
<a class="sourceLine" id="cb10-21" data-line-number="21">                                        <span class="dt">empty_drop_matrix =</span> neg_adt_matrix)</a>
<a class="sourceLine" id="cb10-22" data-line-number="22"></a>
<a class="sourceLine" id="cb10-23" data-line-number="23"></a>
<a class="sourceLine" id="cb10-24" data-line-number="24"><span class="co"># add the assay to the Seurat object </span></a>
<a class="sourceLine" id="cb10-25" data-line-number="25">singlet =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/Seurat/man/SetAssayData.html">SetAssayData</a></span>(<span class="dt">object =</span> singlet, <span class="dt">slot =</span> <span class="st">"CITE"</span>, <span class="dt">new.data =</span> normalized_matrix)</a></code></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/MattPM/dsb">https://​github.com/​MattPM/​dsb</a>
</li>
<li>Report a bug at <br><a href="https://github.com/MattPM/dsb/issues">https://​github.com/​MattPM/​dsb/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li> <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
