<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>dsb_normalizing_CITEseq_data • dsb</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="dsb_normalizing_CITEseq_data">
<meta property="og:description" content="dsb">
<meta property="og:image" content="https://mattpm.github.io/dsb/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dsb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/dsb_normalizing_CITEseq_data.html">dsb_normalizing_CITEseq_data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/MattPM/dsb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>dsb_normalizing_CITEseq_data</h1>
                        <h4 class="author">MPM</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/MattPM/dsb/blob/master/vignettes/dsb_normalizing_CITEseq_data.Rmd"><code>vignettes/dsb_normalizing_CITEseq_data.Rmd</code></a></small>
      <div class="hidden name"><code>dsb_normalizing_CITEseq_data.Rmd</code></div>

    </div>

    
    
<p>Detailed usage of the DSB package for normalizing CITEseq data</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>installation</h2>
<p>You can install the release version of dsb in your R session with the command below</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"># this is analagous to install.packages("package"), you have the package "devtools"  to install a package from a github repository like this one. 



# this is the command you would use to install the package: 
# require(devtools)
# devtools::install_github(repo = 'MattPM/dsb')</pre></body></html></div>
<p><strong>Workflow for usage with Seurat version 3</strong> This is a simple but comprehensive example of loading the output from CITEseq-count, reading in the output from Cellranger (mRNA), merging all the assays together, demultiplexing hashing data to get negative droplets, adding the negative droplets, using the negative droplets to normalize with the DSBNormalizeProtein function, and adding those normalized counts back to an object containing normalized singlets.</p>
<p><strong>Many thanks to Can Liu for testing the DSB package and helping assemble this vignette</strong></p>
</div>
<div id="load-packages-used-in-this-script" class="section level2">
<h2 class="hasAnchor">
<a href="#load-packages-used-in-this-script" class="anchor"></a>Load packages used in this script</h2>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dsb</span>))
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">Seurat</span>))
<span class="co">#&gt; Warning: package 'Seurat' was built under R version 3.6.2</span>
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>))
<span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressMessages</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>))</pre></body></html></div>
<p><strong>Read the output from CITE-seq-count and raw_bc_matrix from cellranger</strong></p>
<p>It is highly recommended that you load the raw_feature_bc_matrix from cellranger, <em>not</em> the filtered_bc_matrix. It is not clear whether the cellranger filtering will accomodate doublets in a superloading (i.e. hashing or multiseq) experiment. More importantly, demultiplexing functions such as HTODemux estimate the background distribution for each cell a k medioids clustering method using k = 1 + n HTO in order to find the negative cells which are then used to establish a threshold for positivity of each HTO. Without sufficient negative droplets (i.e. if you use the filtered_bc_matrix) there may not be enough negative droplets to estimate the background causing the function to not converge. The empty droplets are also crucial for the DSB notmalization step so having more empty droplets makes the protein background estimation more robust.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="co">## Part 1 read in CITE-seq-count output </span>
<span class="co"># Run CITEeq count with expect cells set higher than your actual expected loading </span>
<span class="co"># e.g. if expect 25K cells, set expect cells to 40K </span>
<span class="co"># refer to the CITEseq count documentation page https://hoohm.github.io/CITE-seq-Count/</span>

<span class="co"># read in HTO data (output from citeseq count)</span>
<span class="no">hto_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="st">'citeseq_count/HTO_umi_40k/'</span>, <span class="kw">gene.column</span><span class="kw">=</span><span class="fl">1</span>)


<span class="co"># read in ADT data (output from citeseq count)</span>
<span class="no">adt_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="st">'citeseq_count/ADT_umi_40k/'</span>, <span class="kw">gene.column</span><span class="kw">=</span><span class="fl">1</span>)


<span class="co"># read in mRNA data (output from 10x cellranger)</span>
<span class="no">rna_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Read10X.html">Read10X</a></span>(<span class="kw">data.dir</span> <span class="kw">=</span> <span class="st">"cellranger_count/raw_feature_bc_matrix/"</span>)


<span class="co"># get joint barcodes, subset RNA, HTO, ADT remove "unmapped" row in ADT and HTO assay (if using CITEseq count)</span>
<span class="no">joint.bcs</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span>(<span class="st">"intersect"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">rna_data</span>), <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">hto_data</span>), <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">adt_data</span>)))
<span class="no">rna_data</span> <span class="kw">=</span> <span class="no">rna_data</span>[, <span class="no">joint.bcs</span>]
<span class="no">hto_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">hto_data</span>[-<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">hto_data</span>), <span class="no">joint.bcs</span>])
<span class="no">adt_data</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">adt_data</span>[-<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(<span class="no">adt_data</span>), <span class="no">joint.bcs</span>])</pre></body></html></div>
<p><strong>Create a Seurat object and normalize</strong></p>
<p>we now have 3 separate matrices, one for mRNA , protein and hashing data. Since this was a multiplesing experiment , we also have doublets, singlets and negative (empty) droplets. Here we demultiplex with the HTODemux function to get singlet, doublet and negative annotations for each cell. We then assign the protein data from the negative (empty) droplets and the actual cells (singlets) to variables, and use them in the DSBNormalizeProtein function.</p>
<p>If not using a sample demultiplexing method, we could create a seurat object with a very low threshold for number of genes per cell (i.e. 1), then use a dedicated package that estimates empty droplets (EmptyDrops) or use a threshold, e.g. all droplets with 0-40 genes detected are called as “empty” and are used to estimate the background. (see end of this vignette)</p>
<p>For more information vignettes on demultiplexing: <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html" class="uri">https://satijalab.org/seurat/v3.1/hashing_vignette.html</a> <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq" class="uri">https://github.com/chris-mcginnis-ucsf/MULTI-seq</a></p>
<div class="sourceCode" id="cb4"><html><body><pre class="r">
<span class="co"># create Seurat object using Seurat version 3</span>
<span class="no">SeuratObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateSeuratObject.html">CreateSeuratObject</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">rna_data</span>, <span class="kw">project</span> <span class="kw">=</span> <span class="st">"my_project"</span>)
<span class="no">SeuratObject</span><span class="kw">[[</span><span class="st">"HTO"</span>]] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateAssayObject.html">CreateAssayObject</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">hto_data</span>)
<span class="no">SeuratObject</span><span class="kw">[[</span><span class="st">"ADT"</span>]] <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/CreateAssayObject.html">CreateAssayObject</a></span>(<span class="kw">counts</span> <span class="kw">=</span> <span class="no">adt_data</span>)


<span class="co"># HTO demultiplex -- get the empty droplets from HTO demultiplex</span>
<span class="no">SeuratObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/NormalizeData.html">NormalizeData</a></span>(<span class="no">SeuratObject</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"HTO"</span>, <span class="kw">normalization.method</span> <span class="kw">=</span> <span class="st">"CLR"</span>)
<span class="no">SeuratObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/HTODemux.html">HTODemux</a></span>(<span class="no">SeuratObject</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"HTO"</span>, <span class="kw">positive.quantile</span> <span class="kw">=</span> <span class="fl">0.99</span>)
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">SeuratObject</span>$<span class="no">HTO_classification.global</span>)


<span class="co"># empty droplets from HTO demultiplex as idents = "Negative"</span>
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(<span class="no">SeuratObject</span>) <span class="kw">=</span> <span class="st">"HTO_classification.global"</span>
<span class="no">NegativeObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">SeuratObject</span>, <span class="kw">idents</span> <span class="kw">=</span> <span class="st">"Negative"</span>)
<span class="no">SNG_SeuratObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">SeuratObject</span>, <span class="kw">idents</span> <span class="kw">=</span> <span class="st">"Singlet"</span>)


<span class="co"># before normalizing make sure the negative droplets from hashing are true negatives. </span>
<span class="co"># many ways to do this. Here is a simple heuristic:</span>
<span class="no">NegativeObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">NegativeObject</span>, <span class="kw">subset</span> <span class="kw">=</span> <span class="no">nFeature_RNA</span> <span class="kw">&lt;</span> <span class="fl">40</span>)


<span class="co"># get the empty droplet matrix and singlet matrix</span>
<span class="no">neg_adt_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(<span class="no">NegativeObject</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"ADT"</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">'counts'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()
<span class="no">positive_adt_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(<span class="no">SNG_SeuratObject</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"ADT"</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">'counts'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()


<span class="co"># normalize the data with dsb normalization </span>
<span class="no">isotypes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"isotype_control_1_PROT"</span>, <span class="st">"isotype_control_2_PROT"</span>, <span class="st">"isotype_control_3_PROT"</span>, <span class="st">"isotype_control_4_PROT"</span>)
<span class="no">normalized_matrix</span> <span class="kw">=</span> <span class="fu"><a href="../reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">positive_adt_matrix</span>,
                                        <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">neg_adt_matrix</span>,
                                        <span class="kw">use.isotype.control</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                                        <span class="kw">isotype.control.name.vec</span> <span class="kw">=</span> <span class="no">isotypes</span>)


<span class="co"># add normalized data back to the object (the singlets defined above as "object")</span>
<span class="no">SNG_SeuratObject</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SetAssayData.html">SetAssayData</a></span>(<span class="kw">object</span> <span class="kw">=</span> <span class="no">SNG_SeuratObject</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"ADT"</span>, <span class="kw">new.data</span> <span class="kw">=</span> <span class="no">normalized_matrix</span>)


<span class="co"># proceed with downstream analysis. </span>
<span class="co">#saveRDS(SNG_SeuratObject, file = "my_project_normalized.rds")</span></pre></body></html></div>
</div>
<div id="additional-information-using-the-example-data" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-information-using-the-example-data" class="anchor"></a>Additional information using the example data:</h2>
<p>Here is the format of the required input data.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span>(<span class="no">cells_citeseq_mtx</span>)
<span class="co">#&gt; [1] "matrix"</span>
<span class="no">cells_citeseq_mtx</span>[<span class="fl">27</span>:<span class="fl">37</span>,<span class="fl">1</span>:<span class="fl">2</span>]
<span class="co">#&gt;            GAACGGATCGAATCCA_H1B1ln2 ATAGACCAGTCCATAC_H1B1ln3</span>
<span class="co">#&gt; CD303_PROT                        1                        5</span>
<span class="co">#&gt; CD32_PROT                         6                        9</span>
<span class="co">#&gt; CD357_PROT                        1                        1</span>
<span class="co">#&gt; CD366_PROT                        6                        3</span>
<span class="co">#&gt; CD39_PROT                         4                        6</span>
<span class="co">#&gt; CD40_PROT                         5                        3</span>
<span class="co">#&gt; CD57_PROT                         4                       13</span>
<span class="co">#&gt; CD62L_PROT                       97                      140</span>
<span class="co">#&gt; CD64_PROT                         3                        3</span>
<span class="co">#&gt; CD69_PROT                         8                        9</span>
<span class="co">#&gt; CD70_PROT                         0                        1</span></pre></body></html></div>
<p>You can see above that the input data is a matrix with cells as columns. The non normalized the protein assay of a Seurat object or Single Cell Experiment object is often a sparse matrix. You can easily convert it into a matrix</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co"># </span>
<span class="no">r_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>(<span class="no">sparse_matrix</span>)</pre></body></html></div>
<p>Empty drops are formatted the same way. This is a direct measurement of experimental noise that we correct for with the dsb normalization method.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r">
<span class="no">empty_drop_citeseq_mtx</span>[<span class="fl">27</span>:<span class="fl">37</span>,<span class="fl">1</span>:<span class="fl">2</span>]
<span class="co">#&gt;            AAACCTGCACGGCCAT_H1B1ln1 ACTGCTCAGCAGGCTA_H1B1ln1</span>
<span class="co">#&gt; CD303_PROT                        4                        1</span>
<span class="co">#&gt; CD32_PROT                         8                       14</span>
<span class="co">#&gt; CD357_PROT                        3                        9</span>
<span class="co">#&gt; CD366_PROT                        5                        6</span>
<span class="co">#&gt; CD39_PROT                         4                        6</span>
<span class="co">#&gt; CD40_PROT                         1                        6</span>
<span class="co">#&gt; CD57_PROT                        11                       15</span>
<span class="co">#&gt; CD62L_PROT                        6                        5</span>
<span class="co">#&gt; CD64_PROT                         3                        6</span>
<span class="co">#&gt; CD69_PROT                        10                        6</span>
<span class="co">#&gt; CD70_PROT                         2                        3</span></pre></body></html></div>
</div>
<div id="how-to-get-a-vector-of-isotype-control-proteins" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-get-a-vector-of-isotype-control-proteins" class="anchor"></a>How to get a vector of isotype control proteins</h2>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># protein names in the example data</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">cells_citeseq_mtx</span>)
<span class="co">#&gt;  [1] "AnnexinV_PROT"               "BTLA_PROT"                  </span>
<span class="co">#&gt;  [3] "CD117_PROT"                  "CD123_PROT"                 </span>
<span class="co">#&gt;  [5] "CD13_PROT"                   "CD133_PROT"                 </span>
<span class="co">#&gt;  [7] "CD134_PROT"                  "CD137_PROT"                 </span>
<span class="co">#&gt;  [9] "CD141_PROT"                  "CD152_PROT"                 </span>
<span class="co">#&gt; [11] "CD161_PROT"                  "CD163_PROT"                 </span>
<span class="co">#&gt; [13] "CD18_PROT"                   "CD184_PROT"                 </span>
<span class="co">#&gt; [15] "CD19_PROT"                   "CD1c_PROT"                  </span>
<span class="co">#&gt; [17] "CD1d_PROT"                   "CD2_PROT"                   </span>
<span class="co">#&gt; [19] "CD206_PROT"                  "CD21_PROT"                  </span>
<span class="co">#&gt; [21] "CD223_PROT"                  "CD24_PROT"                  </span>
<span class="co">#&gt; [23] "CD244_PROT"                  "CD273_PROT"                 </span>
<span class="co">#&gt; [25] "CD294_PROT"                  "CD3_PROT"                   </span>
<span class="co">#&gt; [27] "CD303_PROT"                  "CD32_PROT"                  </span>
<span class="co">#&gt; [29] "CD357_PROT"                  "CD366_PROT"                 </span>
<span class="co">#&gt; [31] "CD39_PROT"                   "CD40_PROT"                  </span>
<span class="co">#&gt; [33] "CD57_PROT"                   "CD62L_PROT"                 </span>
<span class="co">#&gt; [35] "CD64_PROT"                   "CD69_PROT"                  </span>
<span class="co">#&gt; [37] "CD70_PROT"                   "CD71_PROT"                  </span>
<span class="co">#&gt; [39] "HLA-ABC_PROT"                "HLA-DR_PROT"                </span>
<span class="co">#&gt; [41] "IgA_PROT"                    "KLRG1_PROT"                 </span>
<span class="co">#&gt; [43] "TCRgd_PROT"                  "CX3CR1_PROT"                </span>
<span class="co">#&gt; [45] "CD10_PROT"                   "CD11c_PROT"                 </span>
<span class="co">#&gt; [47] "CD138_PROT"                  "CD14_PROT"                  </span>
<span class="co">#&gt; [49] "CD16_PROT"                   "CD185_PROT"                 </span>
<span class="co">#&gt; [51] "CD195_PROT"                  "CD196_PROT"                 </span>
<span class="co">#&gt; [53] "CD197_PROT"                  "CD25_PROT"                  </span>
<span class="co">#&gt; [55] "CD27_PROT"                   "CD278 _PROT"                </span>
<span class="co">#&gt; [57] "CD279_PROT"                  "CD31_PROT"                  </span>
<span class="co">#&gt; [59] "CD314 _PROT"                 "CD38_PROT"                  </span>
<span class="co">#&gt; [61] "CD4_PROT"                    "CD45RA_PROT"                </span>
<span class="co">#&gt; [63] "CD56_PROT"                   "CD8_PROT"                   </span>
<span class="co">#&gt; [65] "IgD_PROT"                    "IgM_PROT"                   </span>
<span class="co">#&gt; [67] "MouseIgG1kappaisotype_PROT"  "MouseIgG2akappaisotype_PROT"</span>
<span class="co">#&gt; [69] "Mouse IgG2bkIsotype_PROT"    "RatIgG2bkIsotype_PROT"      </span>
<span class="co">#&gt; [71] "CD103_PROT"                  "CD275_PROT"                 </span>
<span class="co">#&gt; [73] "CD28_PROT"                   "CD45RO_PROT"                </span>
<span class="co">#&gt; [75] "CD5_PROT"                    "CD90_PROT"                  </span>
<span class="co">#&gt; [77] "CD11b_PROT"                  "CD127_PROT"                 </span>
<span class="co">#&gt; [79] "CD194_PROT"                  "CD274_PROT"                 </span>
<span class="co">#&gt; [81] "CD33_PROT"                   "CD7_PROT"                   </span>
<span class="co">#&gt; [83] "CD80_PROT"                   "CD86_PROT"                  </span>
<span class="co">#&gt; [85] "CD183_PROT"                  "CD34_PROT"                  </span>
<span class="co">#&gt; [87] "CD20_PROT"</span>

<span class="co"># the isotype controls are index 67 - 70 in this experiment. </span>
<span class="no">isotypes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">cells_citeseq_mtx</span>)[<span class="fl">67</span>:<span class="fl">70</span>]
<span class="no">isotypes</span>
<span class="co">#&gt; [1] "MouseIgG1kappaisotype_PROT"  "MouseIgG2akappaisotype_PROT"</span>
<span class="co">#&gt; [3] "Mouse IgG2bkIsotype_PROT"    "RatIgG2bkIsotype_PROT"</span></pre></body></html></div>
</div>
<div id="full-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#full-usage" class="anchor"></a>Full usage</h2>
<p>This takes about 8 seconds on a laptop. The method scales well to increased cell numbers (a few minutes to normalize a 50,000 cell dataset) with the slowest part of the script being the per-cell mixture model using the mclust package.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r">
<span class="co"># the isotypes were defined above </span>
<span class="co"># isotypes = c("Mouse IgG2bkIsotype_PROT", "MouseIgG1kappaisotype_PROT","MouseIgG2akappaisotype_PROT", "RatIgG2bkIsotype_PROT")</span>
<span class="no">normalized_matrix</span> <span class="kw">=</span> <span class="fu"><a href="../reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">cells_citeseq_mtx</span>,
                                        <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">empty_drop_citeseq_mtx</span>,
                                        <span class="kw">use.isotype.control</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                                        <span class="kw">isotype.control.name.vec</span> <span class="kw">=</span> <span class="no">isotypes</span>)
<span class="co"># examine the normalized values. </span>
<span class="no">normalized_matrix</span>[<span class="fl">27</span>:<span class="fl">37</span>,<span class="fl">1</span>:<span class="fl">2</span>]
<span class="co">#&gt;            GAACGGATCGAATCCA_H1B1ln2 ATAGACCAGTCCATAC_H1B1ln3</span>
<span class="co">#&gt; CD303_PROT              -1.26428471               1.65042534</span>
<span class="co">#&gt; CD32_PROT               -3.28403477              -0.67852898</span>
<span class="co">#&gt; CD357_PROT              -1.25744634              -0.96019055</span>
<span class="co">#&gt; CD366_PROT              -0.35994530              -0.63616534</span>
<span class="co">#&gt; CD39_PROT               -3.10804059              -0.27371655</span>
<span class="co">#&gt; CD40_PROT               -0.02171297              -0.08506246</span>
<span class="co">#&gt; CD57_PROT               -2.08329200               0.44888796</span>
<span class="co">#&gt; CD62L_PROT              11.10775135              12.84111544</span>
<span class="co">#&gt; CD64_PROT               -3.78263849              -1.65680168</span>
<span class="co">#&gt; CD69_PROT               -0.94835757               0.27614132</span>
<span class="co">#&gt; CD70_PROT               -1.47050086              -0.10164575</span></pre></body></html></div>
</div>
<div id="plot-protieins-on-biaxial-gates" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-protieins-on-biaxial-gates" class="anchor"></a>plot protieins on biaxial gates</h2>
<div class="sourceCode" id="cb10"><html><body><pre class="r">
<span class="co"># to plot the data in ggplot we need to make the proteins the variables (columns)</span>
<span class="no">data.plot</span> <span class="kw">=</span> <span class="no">normalized_matrix</span> <span class="kw">%&gt;%</span>
  <span class="no">t</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="no">CD4_PROT</span>, <span class="no">CD8_PROT</span>, <span class="no">CD27_PROT</span>, <span class="no">CD19_PROT</span>)

<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>(<span class="no">data.plot</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">CD4_PROT</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">CD8_PROT</span>)) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span>(<span class="kw">size</span> <span class="kw">=</span> <span class="fl">0.2</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density_2d.html">geom_density_2d</a></span>(<span class="kw">color</span> <span class="kw">=</span> <span class="st">"black"</span>) +
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span>()</pre></body></html></div>
<p><img src="dsb_normalizing_CITEseq_data_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="integration-with-seurat-also-see-package-readme" class="section level2">
<h2 class="hasAnchor">
<a href="#integration-with-seurat-also-see-package-readme" class="anchor"></a>Integration with Seurat (also see package readme)</h2>
</div>
<div id="how-do-i-get-the-empty-droplets" class="section level2">
<h2 class="hasAnchor">
<a href="#how-do-i-get-the-empty-droplets" class="anchor"></a>How do I get the empty droplets?</h2>
<p>There are a number of ways to get the empty drops. If you are using cell hashing, when you demultiplex the cells, you get a vector of empty or Negative droplets; see the workflow above.</p>
<p>HTODemux function in Seurat: <a href="https://satijalab.org/seurat/v3.1/hashing_vignette.html" class="uri">https://satijalab.org/seurat/v3.1/hashing_vignette.html</a></p>
<p>deMULTIplex function from Multiseq (this is now also implemented in Seurat). <a href="https://github.com/chris-mcginnis-ucsf/MULTI-seq" class="uri">https://github.com/chris-mcginnis-ucsf/MULTI-seq</a></p>
</div>
<div id="simple-example-workflow-seurat-version-3" class="section level2">
<h2 class="hasAnchor">
<a href="#simple-example-workflow-seurat-version-3" class="anchor"></a>Simple example workflow (Seurat Version 3)</h2>
<div class="sourceCode" id="cb11"><html><body><pre class="r">
<span class="co"># get the ADT counts using Seurat version 3 </span>
<span class="no">seurat_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/HTODemux.html">HTODemux</a></span>(<span class="no">seurat_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"HTO"</span>, <span class="kw">positive.quantile</span> <span class="kw">=</span> <span class="fl">0.99</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(<span class="no">seurat_object</span>) <span class="kw">=</span> <span class="st">"HTO_classification.global"</span>
<span class="no">neg_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seurat_object</span>, <span class="kw">idents</span> <span class="kw">=</span> <span class="st">"Negative"</span>)
<span class="no">singlet_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seurat_object</span>, <span class="kw">idents</span> <span class="kw">=</span> <span class="st">"Singlet"</span>)


<span class="co"># non sparse CITEseq data actually store better in a regular materix so the as.matrix() call is not memory intensive.</span>
<span class="no">neg_adt_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(<span class="no">neg_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"CITE"</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">'counts'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()
<span class="no">positive_adt_matrix</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/GetAssayData.html">GetAssayData</a></span>(<span class="no">singlet_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"CITE"</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">'counts'</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()


<span class="co"># normalize the data with dsb</span>
<span class="co"># make sure you've run devtools::install_github(repo = 'MattPM/dsb')</span>
<span class="no">normalized_matrix</span> <span class="kw">=</span> <span class="fu"><a href="../reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">positive_adt_matrix</span>,
                                        <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">neg_adt_matrix</span>)


<span class="co"># now add the normalized dat back to the object (the singlets defined above as "object")</span>
<span class="no">singlet_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SetAssayData.html">SetAssayData</a></span>(<span class="kw">object</span> <span class="kw">=</span> <span class="no">singlet_object</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">"CITE"</span>, <span class="kw">new.data</span> <span class="kw">=</span> <span class="no">normalized_matrix</span>)</pre></body></html></div>
<p>As an additional QC step one can confirm the cells called as “Negative” have low RNA / gene content to be certain there are no contaminating cells.</p>
<p>Also it is not necessary but we reccomend hash demultiplexing with the raw output from cellranger rather than the processed output (i.e. outs/raw_feature_bc_matrix) will have more empty droplets from which the HTODemux function will be able to estimate the negative distribution. This will also have the benefit of creating more droplets to use as built protein background controls in the DSB function.</p>
</div>
<div id="example-workflow-seurat-version-2" class="section level2">
<h2 class="hasAnchor">
<a href="#example-workflow-seurat-version-2" class="anchor"></a>example workflow Seurat version 2</h2>
<div class="sourceCode" id="cb12"><html><body><pre class="r">
<span class="co"># get the ADT counts using Seurat version 3 </span>
<span class="no">seurat_object</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/HTODemux.html">HTODemux</a></span>(<span class="no">seurat_object</span>, <span class="kw">assay</span> <span class="kw">=</span> <span class="st">"HTO"</span>, <span class="kw">positive.quantile</span> <span class="kw">=</span> <span class="fl">0.99</span>)

<span class="no">neg</span> <span class="kw">=</span> <span class="no">seurat_object</span> <span class="kw">%&gt;%</span>
  <span class="fu">SetAllIdent</span>(<span class="kw">id</span> <span class="kw">=</span> <span class="st">"hto_classification_global"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SubsetData.html">SubsetData</a></span>(<span class="kw">ident.use</span> <span class="kw">=</span> <span class="st">"Negative"</span>)

<span class="no">singlet</span> <span class="kw">=</span> <span class="no">seurat_object</span> <span class="kw">%&gt;%</span>
  <span class="fu">SetAllIdent</span>(<span class="kw">id</span> <span class="kw">=</span> <span class="st">"hto_classification_global"</span>) <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SubsetData.html">SubsetData</a></span>(<span class="kw">ident.use</span> <span class="kw">=</span> <span class="st">"Singlet"</span>)

<span class="co"># get negative and positive ADT data </span>
<span class="no">neg_adt_matrix</span> <span class="kw">=</span> <span class="no">neg</span>@<span class="kw">assay</span>$<span class="no">CITE</span>@<span class="kw">raw.data</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()
<span class="no">pos_adt_matrix</span> <span class="kw">=</span> <span class="no">singlet</span>@<span class="kw">assay</span>$<span class="no">CITE</span>@<span class="kw">raw.data</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span>()


<span class="co"># normalize the data with dsb</span>
<span class="co"># make sure you've run devtools::install_github(repo = 'MattPM/dsb')</span>
<span class="no">normalized_matrix</span> <span class="kw">=</span> <span class="fu"><a href="../reference/DSBNormalizeProtein.html">DSBNormalizeProtein</a></span>(<span class="kw">cell_protein_matrix</span> <span class="kw">=</span> <span class="no">pos_adt_matrix</span>,
                                        <span class="kw">empty_drop_matrix</span> <span class="kw">=</span> <span class="no">neg_adt_matrix</span>)


<span class="co"># add the assay to the Seurat object </span>
<span class="no">singlet</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/SetAssayData.html">SetAssayData</a></span>(<span class="kw">object</span> <span class="kw">=</span> <span class="no">singlet</span>, <span class="kw">slot</span> <span class="kw">=</span> <span class="st">"CITE"</span>, <span class="kw">new.data</span> <span class="kw">=</span> <span class="no">normalized_matrix</span>)</pre></body></html></div>
</div>
<div id="how-to-get-empty-drops-without-cell-hashing-or-sample-demultiplexing" class="section level2">
<h2 class="hasAnchor">
<a href="#how-to-get-empty-drops-without-cell-hashing-or-sample-demultiplexing" class="anchor"></a>How to get empty drops without cell hashing or sample demultiplexing?</h2>
<p>If you didn’t run a multiplexing experiment you can simply get a vector of negative droplets from the droplets that ould be QCd out of the experiment due to very low mRNA counts as an estimation of droplets that contain only ambient loading buffer and no cells. There is also an excellent R package for this.<br><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y" class="uri">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y</a></p>
<p>This is why it is critical to use the raw_bc_output from e.g. cellranger rather than using cellrangers filtered ouput with the estimated cells only.</p>
<p>There are 2 parameters you can tune to increase the number of cells recoverd per lane. <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/count</a> –expect-cells (optional) Expected number of recovered cells. Default: 3,000 cells. –force-cells (optional) Force pipeline to use this number of cells, bypassing the cell detection algorithm. Use this if the number of cells estimated by Cell Ranger is not consistent with the barcode rank plot.</p>
<p>Below is a quick method to get outlier empty droplets assuming seurat_object is a object with most cells (i.e. any cell expressing at least a gene).</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="co"># get the nUMI from a seurat version 3 object </span>
<span class="no">umi</span> <span class="kw">=</span> <span class="no">seurat_object</span>$<span class="no">nUMI</span>

<span class="co">#  Get the nUMI from a Seurat version 2 objec </span>
<span class="no">umi</span> <span class="kw">=</span> <span class="no">seurat_object</span>@<span class="kw">meta.data</span> <span class="kw">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(<span class="st">"nUMI"</span>)
<span class="no">mu_umi</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(<span class="no">umi</span>)
<span class="no">sd_umi</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(<span class="no">umi</span>)

<span class="co"># calculate a threshold for calling a cell negative </span>
<span class="no">sub_threshold</span> <span class="kw">=</span> <span class="no">mu_umi</span> - (<span class="fl">5</span>*<span class="no">sd_umi</span>)
<span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/Idents.html">Idents</a></span>(<span class="no">seurat_object</span>) <span class="kw">=</span> <span class="st">"nUMI"</span>

<span class="co"># define the negative cell object</span>
<span class="no">neg</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Seurat/man/subset.Seurat.html">subset</a></span>(<span class="no">seurat_object</span>, <span class="kw">accept.high</span> <span class="kw">=</span> <span class="no">sub_threshold</span>)</pre></body></html></div>
<p>This negative cell object can be used to define the negative background following the examples above.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by .</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
